// devframe/dist/adapters/build.mjs
import { n as strictJsonStringify } from "../serialization-CwO31axa.mjs";
import { i as structuredCloneStringify } from "../structured-clone-DcDc2Dds.mjs";
import { t as colors } from "../colors-Bl4W18Mh.mjs";
import { t as createHostContext } from "../context-DUibxhrR.mjs";
import { t as createH3DevToolsHost } from "../host-h3-Dz8YnFgu.mjs";
import { DEVTOOLS_CONNECTION_META_FILENAME, DEVTOOLS_RPC_DUMP_DIRNAME, DEVTOOLS_RPC_DUMP_MANIFEST_FILENAME } from "../constants.mjs";
import { t as collectStaticRpcDump } from "../static-dump-D7AvjrgL.mjs";
import { n as resolveBasePath } from "../_shared-DcJX9-Az.mjs";
import { existsSync } from "node:fs";
import process from "node:process";
import { dirname, resolve } from "pathe";
import fs$1 from "node:fs/promises";
//#region src/adapters/build.ts
/**
* Produce a self-contained static deploy of a devframe:
*
*   - Build a `mode: 'build'` context and run `devframe.setup(ctx)`.
*   - Copy the author's SPA dist into `<outDir>/`.
*   - Write `<outDir>/__connection.json` (`{ backend: 'static' }`) and the
*     sharded RPC dump under `<outDir>/__rpc-dump/` so the deployed SPA
*     discovers both via relative paths from `document.baseURI`.
*   - When `def.spa` is configured, also write `<outDir>/spa-loader.json`
*     describing the SPA's data-loader mode (`'query'` / `'upload'` /
*     `'none'`). The output is mount-path agnostic — the same bundle
*     works at `/`, `/devtools/`, or any base, no rewriting required.
*/
async function createBuild(d, options = {}) {
	const outDir = resolve(options.outDir ?? "dist-static");
	const distDir = options.distDir ?? d.cli?.distDir;
	if (!distDir) throw new Error(`[devframe] createBuild: no distDir for "${d.id}". Set \`cli.distDir\` on the definition or pass it as an option.`);
	if (existsSync(outDir)) await fs$1.rm(outDir, { recursive: true });
	await fs$1.mkdir(outDir, { recursive: true });
	console.log(colors.cyan`[devframe] copying SPA from ${distDir} -> ${outDir}`);
	await fs$1.cp(distDir, outDir, { recursive: true });
	const ctx = await createHostContext({
		cwd: process.cwd(),
		mode: "build",
		host: createH3DevToolsHost({
			origin: "http://localhost",
			appName: d.id
		})
	});
	await d.setup(ctx);
	await fs$1.mkdir(resolve(outDir, DEVTOOLS_RPC_DUMP_DIRNAME), { recursive: true });
	const jsonSerializableMethods = [];
	for (const def of ctx.rpc.definitions.values()) if (def.jsonSerializable === true) jsonSerializableMethods.push(def.name);
	await fs$1.writeFile(resolve(outDir, DEVTOOLS_CONNECTION_META_FILENAME), JSON.stringify({
		backend: "static",
		jsonSerializableMethods
	}, null, 2), "utf-8");
	console.log(colors.cyan`[devframe] writing RPC dump to ${resolve(outDir, DEVTOOLS_RPC_DUMP_MANIFEST_FILENAME)}`);
	const dump = await collectStaticRpcDump(ctx.rpc.definitions.values(), ctx);
	const indent = options.pretty ? 2 : void 0;
	for (const [filepath, file] of Object.entries(dump.files)) {
		const fullpath = resolve(outDir, filepath);
		await fs$1.mkdir(dirname(fullpath), { recursive: true });
		const text = file.serialization === "structured-clone" ? structuredCloneStringify(file.data) : strictJsonStringify(file.data, file.fnName);
		await fs$1.writeFile(fullpath, file.serialization === "json" && indent != null ? JSON.stringify(JSON.parse(text), null, indent) : text, "utf-8");
	}
	await fs$1.writeFile(resolve(outDir, DEVTOOLS_RPC_DUMP_MANIFEST_FILENAME), JSON.stringify(dump.manifest, null, 2), "utf-8");
	if (d.spa) {
		const base = options.base ?? resolveBasePath(d, "standalone");
		const spaLoader = {
			version: 1,
			mode: d.spa.loader ?? "none",
			base
		};
		await fs$1.writeFile(resolve(outDir, "spa-loader.json"), JSON.stringify(spaLoader, null, 2), "utf-8");
	}
	console.log(colors.green`[devframe] built "${d.id}" -> ${outDir}`);
}
//#endregion
export { createBuild };
