// devframe/dist/adapters/mcp.mjs
import { n as logger } from "../diagnostics-BFhNNmST.mjs";
import { t as createHostContext } from "../context-DUibxhrR.mjs";
import { homedir } from "node:os";
import process from "node:process";
import { join } from "pathe";
import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { CallToolRequestSchema, ListResourcesRequestSchema, ListToolsRequestSchema, ReadResourceRequestSchema } from "@modelcontextprotocol/sdk/types.js";
import { toJsonSchema } from "@valibot/to-json-schema";
//#region src/node/mcp/to-json-schema.ts
const FALLBACK_OBJECT_SCHEMA = Object.freeze({
	type: "object",
	additionalProperties: true
});
/**
* Convert a valibot return schema to JSON Schema.
* @internal
*/
function valibotReturnToJsonSchema(schema) {
	if (!schema) return void 0;
	try {
		return toJsonSchema(schema);
	} catch {
		return FALLBACK_OBJECT_SCHEMA;
	}
}
/**
* Convert positional RPC args schemas to a single MCP-friendly object
* schema. When the RPC declares `args: [v.object(...)]`, unwrap the
* single-object schema directly (nicer agent UX than `{ arg0: {...} }`).
*
* Returns `undefined` when there are no args (the MCP SDK treats this
* as `{ type: 'object', properties: {} }`).
* @internal
*/
function valibotArgsToJsonSchema(args) {
	if (!args || args.length === 0) return {
		schema: {
			type: "object",
			properties: {}
		},
		unwrapped: false
	};
	if (args.length === 1) {
		const inner = safeToJsonSchema(args[0]);
		if (isObjectJsonSchema(inner)) return {
			schema: inner,
			unwrapped: true
		};
	}
	const properties = {};
	const required = [];
	for (let i = 0; i < args.length; i++) {
		const key = `arg${i}`;
		properties[key] = safeToJsonSchema(args[i]);
		required.push(key);
	}
	return {
		schema: {
			type: "object",
			properties,
			required,
			additionalProperties: false
		},
		unwrapped: false
	};
}
function safeToJsonSchema(schema) {
	try {
		return toJsonSchema(schema);
	} catch {
		return FALLBACK_OBJECT_SCHEMA;
	}
}
function isObjectJsonSchema(value) {
	return !!value && typeof value === "object" && value.type === "object";
}
//#endregion
//#region src/node/mcp/build-server.ts
/**
* Wire an MCP {@link Server} to a devframe context. Returns the server
* plus a disposal function for the subscriptions it sets up. The
* transport is the caller's responsibility — `createMcpServer` connects
* stdio; tests can connect an {@link InMemoryTransport} instead.
*
* @internal
*/
function buildMcpServerFromContext(ctx, options) {
	const server = new Server({
		name: options.serverName,
		version: options.serverVersion
	}, { capabilities: {
		tools: { listChanged: true },
		resources: { listChanged: true }
	} });
	registerToolHandlers(server, ctx);
	registerResourceHandlers(server, ctx, options.exposeSharedState);
	const notify = (method) => {
		server.notification({ method }).catch(() => {});
	};
	const offManifest = ctx.agent.events.on("agent:manifest:changed", () => {
		notify("notifications/tools/list_changed");
		notify("notifications/resources/list_changed");
	});
	const offKeyAdded = ctx.rpc.sharedState.onKeyAdded(() => {
		notify("notifications/resources/list_changed");
	});
	return {
		server,
		dispose: () => {
			offManifest();
			offKeyAdded();
		}
	};
}
/**
* Build an MCP server over the agent surface of a devframe definition.
* Currently supports `stdio` transport only.
*
* @experimental The agent-native surface is experimental and may change
* without a major version bump until it stabilizes.
*/
async function createMcpServer(definition, options = {}) {
	const transport = options.transport ?? "stdio";
	if (transport !== "stdio") throw logger.DF0017({
		transport,
		reason: "Only stdio transport is supported in this release."
	}).throw();
	const ctx = await createHostContext({
		cwd: process.cwd(),
		mode: "dev",
		host: {
			mountStatic: () => {},
			resolveOrigin: () => "mcp://devframe",
			getStorageDir: (scope) => scope === "workspace" ? join(process.cwd(), `node_modules/.${definition.id}/devtools`) : join(homedir(), `.${definition.id}/devtools`)
		}
	});
	await definition.setup(ctx);
	const { server, dispose } = buildMcpServerFromContext(ctx, {
		serverName: options.serverName ?? `${definition.id} (devframe)`,
		serverVersion: options.serverVersion ?? definition.version ?? "0.0.0",
		exposeSharedState: options.exposeSharedState ?? true
	});
	const { startStdioTransport } = await import("../transports-BnCs4rj9.mjs");
	let stop;
	try {
		stop = await startStdioTransport(server);
	} catch (error) {
		const reason = error instanceof Error ? error.message : String(error);
		throw logger.DF0017({
			transport,
			reason
		}, { cause: error }).throw();
	}
	options.onReady?.({ transport: "stdio" });
	return { async stop() {
		dispose();
		await stop();
	} };
}
function registerToolHandlers(server, ctx) {
	server.setRequestHandler(ListToolsRequestSchema, async () => {
		return { tools: ctx.agent.list().tools.map((tool) => projectTool(tool, ctx)) };
	});
	server.setRequestHandler(CallToolRequestSchema, async (request) => {
		const { name, arguments: args } = request.params;
		try {
			return { content: [{
				type: "text",
				text: stringify(await ctx.agent.invoke(name, args ?? {}))
			}] };
		} catch (error) {
			return {
				isError: true,
				content: [{
					type: "text",
					text: `Error invoking "${name}": ${error instanceof Error ? error.message : String(error)}`
				}]
			};
		}
	});
}
function registerResourceHandlers(server, ctx, exposeSharedState) {
	server.setRequestHandler(ListResourcesRequestSchema, async () => {
		const resources = ctx.agent.list().resources.map((resource) => ({
			uri: resource.uri,
			name: resource.name,
			description: resource.description,
			mimeType: resource.mimeType
		}));
		if (exposeSharedState !== false) {
			const filter = typeof exposeSharedState === "function" ? exposeSharedState : () => true;
			for (const key of ctx.rpc.sharedState.keys()) {
				if (!filter(key)) continue;
				resources.push({
					uri: `devframe://state/${encodeURIComponent(key)}`,
					name: key,
					description: `Shared state: ${key}`,
					mimeType: "application/json"
				});
			}
		}
		return { resources };
	});
	server.setRequestHandler(ReadResourceRequestSchema, async (request) => {
		const { uri } = request.params;
		const parsed = parseResourceUri(uri);
		if (parsed.kind === "resource") {
			const content = await ctx.agent.read(parsed.id);
			return { contents: [{
				uri,
				mimeType: content.mimeType ?? "application/json",
				text: content.text ?? stringify(content.json)
			}] };
		}
		if (parsed.kind === "state") return { contents: [{
			uri,
			mimeType: "application/json",
			text: stringify((await ctx.rpc.sharedState.get(parsed.key)).value())
		}] };
		throw new Error(`[devframe/mcp] unknown resource URI "${uri}"`);
	});
}
function projectTool(tool, ctx) {
	const inputSchema = tool.inputSchema ?? computeInputSchema(tool, ctx);
	const outputSchema = tool.outputSchema ?? computeOutputSchema(tool, ctx);
	return {
		name: tool.id,
		title: tool.title,
		description: tool.description,
		inputSchema,
		...outputSchema ? { outputSchema } : {},
		annotations: {
			title: tool.title,
			readOnlyHint: tool.safety === "read",
			destructiveHint: tool.safety === "destructive"
		}
	};
}
function computeInputSchema(tool, ctx) {
	if (tool.kind !== "rpc" || !tool.rpcName) return {
		type: "object",
		properties: {}
	};
	const def = ctx.rpc.definitions.get(tool.rpcName);
	if (!def) return {
		type: "object",
		properties: {}
	};
	const args = def.args;
	return valibotArgsToJsonSchema(args).schema;
}
function computeOutputSchema(tool, ctx) {
	if (tool.kind !== "rpc" || !tool.rpcName) return void 0;
	const def = ctx.rpc.definitions.get(tool.rpcName);
	if (!def) return void 0;
	return valibotReturnToJsonSchema(def.returns);
}
function parseResourceUri(uri) {
	const match = uri.match(/^devframe:\/\/(resource|state)\/(.+)$/);
	if (!match) return { kind: "unknown" };
	const [, kind, rest] = match;
	const decoded = decodeURIComponent(rest);
	if (kind === "resource") return {
		kind: "resource",
		id: decoded
	};
	return {
		kind: "state",
		key: decoded
	};
}
function stringify(value) {
	if (value === void 0) return "undefined";
	if (typeof value === "string") return value;
	try {
		return JSON.stringify(value, null, 2);
	} catch {
		return String(value);
	}
}
//#endregion
export { createMcpServer };
