// devframe/dist/node/index.mjs
import { a as createRpcSharedStateServerHost, i as createRpcStreamingServerHost, n as DevToolsViewHost, o as DevToolsDiagnosticsHost, r as RpcFunctionsHost, s as DevToolsAgentHost, t as createHostContext } from "../context-DUibxhrR.mjs";
import { t as createH3DevToolsHost } from "../host-h3-Dz8YnFgu.mjs";
import { t as startHttpAndWs } from "../server-ma8-ZVyX.mjs";
import { t as collectStaticRpcDump } from "../static-dump-D7AvjrgL.mjs";
import { t as createStorage } from "../storage-CNC38eU-.mjs";
import { isIP } from "node:net";
//#region src/node/utils.ts
function isObject(value) {
	return Object.prototype.toString.call(value) === "[object Object]";
}
function normalizeHttpServerUrl(host, port) {
	return `http://${host === "127.0.0.1" ? "localhost" : isIP(host) === 6 ? `[${host}]` : host}:${port}`;
}
//#endregion
export { DevToolsAgentHost, DevToolsDiagnosticsHost, DevToolsViewHost, RpcFunctionsHost, collectStaticRpcDump, createH3DevToolsHost, createHostContext, createRpcSharedStateServerHost, createRpcStreamingServerHost, createStorage, isObject, normalizeHttpServerUrl, startHttpAndWs };
