// devframe/dist/utils/serve-static.mjs
import { createReadStream } from "node:fs";
import { extname, join, normalize, resolve, sep } from "pathe";
import { defineEventHandler, sendStream, setResponseHeader, setResponseStatus } from "h3";
import { stat } from "node:fs/promises";
import { lookup } from "mrmime";
//#region src/utils/serve-static.ts
const HTML_EXTENSIONS = [".html", ".htm"];
async function statFile(abs) {
	try {
		const s = await stat(abs);
		if (!s.isFile()) return null;
		return {
			abs,
			size: s.size,
			mtime: s.mtime
		};
	} catch {
		return null;
	}
}
async function resolveTarget(absDir, urlPath, indexNames, single) {
	let cleaned;
	try {
		cleaned = decodeURIComponent(urlPath || "/");
	} catch {
		return null;
	}
	cleaned = cleaned.replace(/[?#].*$/, "");
	if (cleaned.endsWith("/")) cleaned = cleaned.slice(0, -1);
	if (cleaned.startsWith("/")) cleaned = cleaned.slice(1);
	const abs = normalize(join(absDir, cleaned));
	if (abs !== absDir && !abs.startsWith(absDir + sep)) return null;
	const direct = await statFile(abs);
	if (direct) return direct;
	try {
		if ((await stat(abs)).isDirectory()) for (const name of indexNames) {
			const candidate = await statFile(join(abs, name));
			if (candidate) return candidate;
		}
	} catch {}
	if (!extname(cleaned)) for (const ext of HTML_EXTENSIONS) {
		const candidate = await statFile(abs + ext);
		if (candidate) return candidate;
	}
	const fallbackIndex = indexNames[0];
	if (single && fallbackIndex && !/\.[a-z0-9]+$/i.test(cleaned)) {
		const indexFile = await statFile(join(absDir, fallbackIndex));
		if (indexFile) return indexFile;
	}
	return null;
}
function contentTypeFor(abs) {
	const type = lookup(abs);
	if (!type) return "application/octet-stream";
	if (type === "text/html") return "text/html; charset=utf-8";
	return type;
}
function setStaticHeaders(res, file) {
	res.setHeader("Content-Type", contentTypeFor(file.abs));
	res.setHeader("Content-Length", file.size);
	res.setHeader("Last-Modified", file.mtime.toUTCString());
	res.setHeader("Cache-Control", "no-store");
}
function normalizeOptions(options) {
	return {
		indexNames: options?.indexNames ?? ["index.html"],
		single: options?.single ?? true
	};
}
/**
* h3 event handler that serves files from `dir` with SPA fallback.
*
* Drop-in replacement for `fromNodeMiddleware(sirv(dir, { dev: true, single: true }))`
* when the surrounding server is an h3 app — no `Cache-Control` beyond
* `no-store`, `Content-Type` resolved via `mrmime`, and a miss with no
* file extension falls back to `<dir>/index.html` so client-side routing
* works.
*/
function serveStaticHandler(dir, options) {
	const absDir = resolve(dir);
	const opts = normalizeOptions(options);
	return defineEventHandler(async (event) => {
		const method = event.node.req.method;
		if (method !== "GET" && method !== "HEAD") {
			setResponseStatus(event, 405);
			setResponseHeader(event, "Allow", "GET, HEAD");
			return "";
		}
		const file = await resolveTarget(absDir, event.node.req.url ?? "/", opts.indexNames, opts.single);
		if (!file) {
			setResponseStatus(event, 404);
			return "";
		}
		setStaticHeaders(event.node.res, file);
		if (method === "HEAD") {
			event.node.res.end();
			return "";
		}
		return sendStream(event, createReadStream(file.abs));
	});
}
/**
* Connect/Express-style Node middleware variant of {@link serveStaticHandler}.
*
* Use when mounting onto `viteServer.middlewares.use(base, …)` or any other
* Connect stack — avoids forcing the host package to depend on h3 just to
* adapt an event handler back into Node middleware.
*/
function serveStaticNodeMiddleware(dir, options) {
	const absDir = resolve(dir);
	const opts = normalizeOptions(options);
	return (req, res, next) => {
		(async () => {
			const method = req.method;
			if (method !== "GET" && method !== "HEAD") {
				if (next) {
					next();
					return;
				}
				res.statusCode = 405;
				res.setHeader("Allow", "GET, HEAD");
				res.end();
				return;
			}
			const file = await resolveTarget(absDir, req.url ?? "/", opts.indexNames, opts.single);
			if (!file) {
				if (next) {
					next();
					return;
				}
				res.statusCode = 404;
				res.end();
				return;
			}
			setStaticHeaders(res, file);
			if (method === "HEAD") {
				res.end();
				return;
			}
			createReadStream(file.abs).pipe(res);
		})().catch((err) => {
			if (next) {
				next(err instanceof Error ? err : new Error(String(err)));
				return;
			}
			res.statusCode = 500;
			res.end();
		});
	};
}
//#endregion
export { serveStaticHandler, serveStaticNodeMiddleware };
