// vue-virtual-scroller/dist/DynamicScroller.vue_vue_type_script_setup_true_lang-BZfgBxGs.js
import { defineComponent as D, ref as F, toRef as E, computed as d, openBlock as O, createBlock as j, mergeProps as x, unref as n, createSlots as A, withCtx as o, renderSlot as s, normalizeProps as U, guardReactiveProps as q } from "vue";
import { useDynamicScroller as G } from "./composables/useDynamicScroller.js";
import { _ as H } from "./RecycleScroller.vue_vue_type_script_setup_true_lang-uvhLi9Vp.js";
import "./scroller.css_vue_type_style_index_0_src_true_lang--7gJ3zh_.js";
const W = /* @__PURE__ */ D({
  inheritAttrs: !1,
  __name: "DynamicScroller",
  props: {
    items: {},
    keyField: { default: "id" },
    direction: { default: "vertical" },
    listTag: { default: "div" },
    itemTag: { default: "div" },
    minItemSize: {},
    shift: { type: Boolean, default: !1 },
    cache: { default: void 0 },
    disableTransform: { type: Boolean, default: !1 },
    flowMode: { type: Boolean, default: !1 },
    hiddenPosition: { default: void 0 },
    enabled: { type: Boolean, default: !0 }
  },
  emits: ["resize", "visible"],
  setup(m, { expose: c, emit: f }) {
    const e = m, r = f, a = F(), p = E(e, "items"), u = d(() => {
      var t;
      const i = (t = a.value) == null ? void 0 : t.el;
      return i && typeof i == "object" && "value" in i ? i.value : i;
    }), h = d(() => ({
      items: p,
      keyField: e.keyField,
      direction: e.direction,
      minItemSize: e.minItemSize,
      shift: e.shift,
      cache: e.cache,
      flowMode: e.flowMode,
      hiddenPosition: e.hiddenPosition,
      enabled: e.enabled,
      el: u.value,
      onResize: () => r("resize"),
      onVisible: () => r("visible")
    })), b = G(
      h
    ), {
      itemsWithSize: y,
      forceUpdate: S,
      startSpacerSize: z,
      endSpacerSize: g,
      scrollToItem: v,
      scrollToPosition: k,
      findItemIndex: I,
      getItemOffset: T,
      getItemSize: B,
      cacheSnapshot: P,
      restoreCache: $,
      scrollToBottom: w,
      onScrollerResize: R,
      onScrollerVisible: _
    } = b;
    function M(i, t, l) {
      return {
        item: i.item,
        index: t,
        active: l,
        itemWithSize: i
      };
    }
    return c({
      startSpacerSize: z,
      endSpacerSize: g,
      scrollToItem: v,
      scrollToPosition: k,
      findItemIndex: I,
      getItemOffset: T,
      scrollToBottom: w,
      getItemSize: B,
      cacheSnapshot: P,
      restoreCache: $,
      forceUpdate: S
    }), (i, t) => (O(), j(H, x({
      ref_key: "scroller",
      ref: a,
      items: n(y),
      "min-item-size": e.minItemSize,
      direction: e.direction,
      cache: e.cache,
      "disable-transform": e.disableTransform,
      "flow-mode": e.flowMode,
      "hidden-position": e.hiddenPosition,
      enabled: e.enabled,
      "key-field": "id",
      "list-tag": e.listTag,
      "item-tag": e.itemTag
    }, i.$attrs, {
      onResize: n(R),
      onVisible: n(_)
    }), A({
      default: o(({ item: l, index: V, active: C }) => [
        s(i.$slots, "default", U(q(M(l, V, C))))
      ]),
      empty: o(() => [
        s(i.$slots, "empty")
      ]),
      _: 2
    }, [
      i.$slots.before ? {
        name: "before",
        fn: o(() => [
          s(i.$slots, "before")
        ]),
        key: "0"
      } : void 0,
      i.$slots.after ? {
        name: "after",
        fn: o(() => [
          s(i.$slots, "after")
        ]),
        key: "1"
      } : void 0
    ]), 1040, ["items", "min-item-size", "direction", "cache", "disable-transform", "flow-mode", "hidden-position", "enabled", "list-tag", "item-tag", "onResize", "onVisible"]));
  }
});
export {
  W as _
};
