// vue-virtual-scroller/dist/useRecycleScroller-G2SVHG1u.js
import { computed as U, toValue as re, ref as fe, onMounted as hn, nextTick as tt, onActivated as vn, onBeforeUnmount as pn, watch as B, markRaw as In, shallowReactive as Sn } from "vue";
import { r as Oe } from "./keyField-D6okV6nC.js";
function ot(n, o, u) {
  return Oe(n, o, u);
}
function Ae(n, o) {
  return n.map((u, c) => ot(u, c, o));
}
function gn(n, o, u) {
  const c = [], l = [];
  for (let d = 0; d < n.length; d++) {
    const y = n[d], x = ot(y, d, o), z = u(y, d, x);
    c.push(x), l.push(typeof z == "number" && z > 0 ? z : null);
  }
  return {
    keys: c,
    sizes: l
  };
}
function yn(n, o, u) {
  if (!n || n.keys.length !== o.length || n.sizes.length !== o.length)
    return !1;
  for (let c = 0; c < o.length; c++)
    if (n.keys[c] !== ot(o[c], c, u))
      return !1;
  return !0;
}
function wn(n, o, u) {
  if (!yn(n, o, u))
    return {};
  const c = {};
  for (let l = 0; l < n.keys.length; l++) {
    const d = n.sizes[l];
    typeof d == "number" && d > 0 && (c[n.keys[l]] = d);
  }
  return c;
}
function xn(n, o) {
  if (!n.length || o.length <= n.length)
    return 0;
  const u = n[0], c = o.indexOf(u);
  if (c <= 0 || c + n.length < o.length && n.length > o.length - c)
    return 0;
  for (let l = 0; l < n.length; l++)
    if (o[c + l] !== n[l])
      return 0;
  return c;
}
function Wt(n, o, u, c, l, d = 0) {
  const y = l ?? "start";
  if (y === "nearest") {
    const x = u + c, z = n + o;
    return n >= u && z <= x ? null : n < u ? n + d : z - c + d;
  }
  return y === "end" ? n + o - c + d : y === "center" ? n + (o - c) / 2 + d : n + d;
}
function bn(n) {
  return n.flowMode && n.direction === "vertical" && !n.gridItems ? "flow" : n.disableTransform ? "position" : "transform";
}
function zn(n, o) {
  const u = o.direction === "vertical", c = o.mode ?? "transform", l = {
    visibility: n.nr.used ? "visible" : "hidden",
    pointerEvents: n.nr.used ? void 0 : "none"
  };
  if (c === "flow" ? l.display = n.nr.used ? void 0 : "none" : (l.position = "absolute", l.top = "0px", l.left = "0px", l.display = void 0), c === "position" ? (l[u ? "top" : "left"] = `${n.position}px`, l[u ? "left" : "top"] = `${n.offset}px`, l.transform = "none", l.willChange = "unset") : c === "transform" && (l.transform = u ? `translateY(${n.position}px) translateX(${n.offset}px)` : `translateX(${n.position}px) translateY(${n.offset}px)`, l.willChange = "transform"), o.gridItems && o.itemSize != null) {
    const d = o.itemSecondarySize || o.itemSize;
    l.width = `${u ? d : o.itemSize}px`, l.height = `${u ? o.itemSize : d}px`;
  }
  return l;
}
function Ot(n) {
  return U(() => re(n));
}
function Mn(n, o, u, c, l) {
  const d = Ot(n);
  return {
    el: U(() => {
      const y = d.value.el;
      return re(o ?? y);
    }),
    before: U(() => {
      const y = d.value.before;
      return re(u ?? y);
    }),
    after: U(() => {
      const y = d.value.after;
      return re(c ?? y);
    }),
    callbacks: {
      onResize: () => {
        var y;
        return (y = (l == null ? void 0 : l.onResize) ?? d.value.onResize) == null ? void 0 : y();
      },
      onVisible: () => {
        var y;
        return (y = (l == null ? void 0 : l.onVisible) ?? d.value.onVisible) == null ? void 0 : y();
      },
      onHidden: () => {
        var y;
        return (y = (l == null ? void 0 : l.onHidden) ?? d.value.onHidden) == null ? void 0 : y();
      },
      onUpdate: (y, x, z, R) => {
        var me;
        return (me = (l == null ? void 0 : l.onUpdate) ?? d.value.onUpdate) == null ? void 0 : me(y, x, z, R);
      }
    }
  };
}
const Tn = {
  itemsLimit: 1e3
};
function Bt(n) {
  return typeof window < "u" && n === window;
}
const Vn = (() => {
  if (typeof document > "u")
    return "negative";
  const n = document.createElement("div"), o = document.createElement("div");
  n.style.width = "4px", n.style.height = "1px", n.style.overflow = "auto", n.style.direction = "rtl", o.style.width = "8px", o.style.height = "1px", n.appendChild(o), document.body.appendChild(n), n.scrollLeft = -1;
  const u = n.scrollLeft < 0;
  return document.body.removeChild(n), u ? "negative" : "default";
})();
function de(n, o, u) {
  return o !== "horizontal" || !u || Bt(u) || getComputedStyle(u).direction !== "rtl" ? n : Vn === "negative" ? -n : n;
}
function En(n, o, u) {
  return de(n, o, u);
}
function nt(n, o, u, c) {
  const l = En(u, o, n), d = !!(c != null && c.smooth);
  if (Bt(n)) {
    o === "vertical" ? n.scrollTo({
      top: l,
      behavior: d ? "smooth" : "auto"
    }) : n.scrollTo({
      left: l,
      behavior: d ? "smooth" : "auto"
    });
    return;
  }
  if (typeof n.scrollTo == "function") {
    n.scrollTo(o === "vertical" ? { top: l, behavior: d ? "smooth" : "auto" } : { left: l, behavior: d ? "smooth" : "auto" });
    return;
  }
  o === "vertical" ? n.scrollTop = l : n.scrollLeft = l;
}
function _n(n, o, u) {
  return u ? o === "vertical" ? window.innerHeight : window.innerWidth : o === "vertical" ? n.clientHeight : n.clientWidth;
}
const Fn = /auto|scroll/;
function Ut(n, o) {
  return n.parentNode === null ? o : Ut(n.parentNode, [...o, n]);
}
function rt(n, o) {
  return getComputedStyle(n, null).getPropertyValue(o);
}
function Pn(n) {
  return rt(n, "overflow") + rt(n, "overflow-y") + rt(n, "overflow-x");
}
function Ln(n) {
  return Fn.test(Pn(n));
}
function He(n) {
  if (!(n instanceof HTMLElement || n instanceof SVGElement))
    return;
  const o = Ut(n.parentNode, []);
  for (let u = 0; u < o.length; u += 1)
    if (o[u] instanceof Element && Ln(o[u]))
      return o[u];
  return document.scrollingElement || document.documentElement;
}
let it = !1;
function Cn() {
  return it;
}
if (typeof window < "u") {
  it = !1;
  try {
    const n = Object.defineProperty({}, "passive", {
      get() {
        it = !0;
      }
    });
    window.addEventListener("test", null, n);
  } catch {
  }
}
function j(n) {
  return typeof n == "number" ? n : null;
}
function Nn(n, o, u, c, l) {
  return typeof u == "number" ? u : typeof u == "function" ? u(n, o) || c || void 0 : c || (n == null ? void 0 : n[l ?? "size"]) || void 0;
}
function Rn(n, o, u, c, l, d) {
  return typeof u == "function" ? u(n, o) || c || Number(l) || 0 : c || (n == null ? void 0 : n[d ?? "size"]) || Number(l) || 0;
}
let kn = 0;
const Wn = [];
function An(n) {
  for (let o = 0; o < n.length; o++)
    n[o];
}
function Hn(n, o) {
  return n && typeof n == "object" ? n[o] : void 0;
}
function Ye(n, o) {
  return n.map((u) => Hn(u, o));
}
function Yn(n, o, u, c) {
  if (n.length !== u.length || o.length !== c.length)
    return !1;
  for (let l = 0; l < n.length; l++)
    if (n[l] !== u[l] || o[l] !== c[l])
      return !1;
  return !0;
}
function On(n) {
  const o = n;
  o._vs_styleStamp++;
}
function At(n) {
  const o = n;
  o._vs_visibilityStamp++, o._vs_styleStamp++;
}
function Bn(n, o, u, c) {
  if (u) {
    for (const l of u)
      c(l);
    return;
  }
  for (let l = n; l < o; l++)
    c(l);
}
const Ht = 8, Yt = 1;
function jn(n, o, u, c, l, d) {
  const y = Ot(n), x = Mn(y, o, u, c, l), z = U(() => {
    const e = re(m().items);
    return An(e), e;
  }), R = fe([]), me = fe(0), he = fe(0), ve = fe(0), ie = fe(!1);
  let D = 0, G = 0, lt = 0, st = 0, ut = !1;
  const Be = /* @__PURE__ */ new Map(), Ue = /* @__PURE__ */ new Map();
  let ze = !1, $e = 0, at = 0, je = !1, oe = null, le = null, X = null, ct = 0, pe = null, Ie = null, Me = [], Se = [], J = null, H = null, ge = null, De = !1, ft = !1, dt = !1;
  const Te = /* @__PURE__ */ new Set(), Ve = fe({}), Ge = [], mt = {
    accumulator: 0
  };
  function m() {
    return y.value;
  }
  function ye() {
    return (d == null ? void 0 : d.pageMode) ?? m().pageMode;
  }
  const M = U(() => re(m().enabled ?? !0)), q = U(() => {
    const e = z.value;
    return e.length > 0 && typeof e[0] != "object";
  }), we = U(() => {
    const e = m();
    if (j(e.itemSize) === null) {
      const r = z.value, i = Array.from({ length: r.length });
      mt.accumulator = 0, i[-1] = mt;
      const s = e.minItemSize, a = Ve.value, h = q.value;
      let g = 1e4, S = 0, T;
      for (let f = 0, v = r.length; f < v; f++) {
        const C = h ? f : Oe(r[f], f, e.keyField);
        T = Rn(
          r[f],
          f,
          e.itemSize,
          a[C],
          s,
          e.sizeField
        ), T < g && (g = T), S += T;
        const k = Ge[f] ?? (Ge[f] = {
          accumulator: 0,
          size: void 0
        });
        k.accumulator = S, k.size = T, i[f] = k;
      }
      return Ge.length = r.length, ct = g, i;
    }
    return Wn;
  }), $t = U(
    () => R.value.filter((e) => (e._vs_visibilityStamp, e.nr.used)).sort((e, t) => e.nr.index - t.nr.index)
  ), jt = U(() => {
    const e = m(), t = q.value ? null : e.keyField;
    return gn(z.value, t, (r, i, s) => Nn(
      r,
      i,
      e.itemSize,
      Ve.value[s],
      e.sizeField
    ));
  });
  function Z(e = m()) {
    return e.direction ?? "vertical";
  }
  function se(e = m()) {
    const t = Z(e);
    let r = e.flowMode ?? !1;
    return r && t !== "vertical" && (ft || (console.warn("[vue-recycle-scroller] flowMode only supports vertical lists. Falling back to standard positioning."), ft = !0), r = !1), r && e.gridItems && (dt || (console.warn("[vue-recycle-scroller] flowMode does not support gridItems. Falling back to standard positioning."), dt = !0), r = !1), bn({
      direction: t,
      disableTransform: e.disableTransform ?? !1,
      flowMode: r,
      gridItems: e.gridItems
    });
  }
  function ht() {
    R.value.sort((e, t) => e.nr.used !== t.nr.used ? e.nr.used ? -1 : 1 : e.nr.used && t.nr.used ? e.nr.index - t.nr.index : e.nr.id - t.nr.id);
  }
  function Xe(e) {
    const t = m();
    return Ve.value = wn(e, z.value, q.value ? null : t.keyField), Object.keys(Ve.value).length > 0;
  }
  function vt(e) {
    let t = Ue.get(e);
    return t || (t = [], Ue.set(e, t)), t;
  }
  function Dt(e, t, r, i, s) {
    const a = In({
      id: kn++,
      index: t,
      used: !0,
      key: i,
      type: s
    }), h = Sn({
      item: r,
      position: 0,
      offset: 0,
      nr: a,
      _vs_styleStamp: 0,
      _vs_visibilityStamp: 0
    });
    return e.push(h), h;
  }
  function Gt(e) {
    const t = vt(e);
    if (t && t.length) {
      const r = t.pop();
      return r.nr.used = !0, At(r), r;
    }
  }
  function pt(e, t) {
    const r = R.value, i = r.indexOf(e);
    if (i === -1)
      return;
    const s = Math.max(0, Math.min(t, r.length - 1));
    i !== s && (r.splice(i, 1), r.splice(s, 0, e));
  }
  function Xt(e) {
    pt(e, R.value.length - 1);
  }
  function qt() {
    const e = R.value;
    let t = 0;
    for (; t < e.length && !e[t].nr.used; )
      t++;
    let r = 0;
    for (let i = t; i < e.length && e[i].nr.used; i++)
      r++;
    return {
      activeStart: t,
      activeCount: r,
      headInsertCount: 0
    };
  }
  function Zt(e, t, r) {
    let i = null;
    t < D ? (i = r.activeStart + r.headInsertCount, r.headInsertCount++, r.activeCount++) : t >= G && (i = r.activeStart + r.activeCount, r.activeCount++), i != null && pt(e, i);
  }
  function It(e, t = !1) {
    const r = e.nr.type;
    vt(r).push(e), e.nr.used = !1, e.position = m().hiddenPosition ?? -999999, At(e), Be.delete(e.nr.key), t && Xt(e);
  }
  function Jt() {
    Be.clear(), Ue.clear();
    for (let e = 0, t = R.value.length; e < t; e++) {
      const r = R.value[e];
      r && It(r);
    }
  }
  function qe(e) {
    let t = -1;
    return t = requestAnimationFrame(() => {
      Te.delete(t), e();
    }), Te.add(t), t;
  }
  function St() {
    for (const e of Te)
      cancelAnimationFrame(e);
    Te.clear();
  }
  function gt() {
    oe && (clearTimeout(oe), oe = null), le && (clearTimeout(le), le = null), X && (clearTimeout(X), X = null), H && (clearTimeout(H), H = null), ge && (clearTimeout(ge), ge = null);
  }
  function Ze() {
    var e, t;
    M.value && ((t = (e = x.callbacks).onResize) == null || t.call(e), ie.value && _(!1));
  }
  function Je() {
    if (!M.value)
      return;
    J && !De && Q();
    const e = m();
    if (!ze) {
      if (ze = !0, oe)
        return;
      const t = () => qe(() => {
        ze = !1;
        const { continuous: r } = _(!1, !0);
        r || (le && clearTimeout(le), le = setTimeout(Je, e.updateInterval + 100));
      });
      t(), e.updateInterval && (oe = setTimeout(() => {
        oe = null, ze && t();
      }, e.updateInterval));
    }
  }
  function Qt(e, t) {
    var r, i, s, a;
    M.value && ie.value && (e || t.boundingClientRect.width !== 0 || t.boundingClientRect.height !== 0 ? ((i = (r = x.callbacks).onVisible) == null || i.call(r), qe(() => {
      _(!1);
    })) : (a = (s = x.callbacks).onHidden) == null || a.call(s));
  }
  function Kt() {
    const e = x.el.value, t = e ? He(e) : void 0;
    return window.document && (t === window.document.documentElement || t === window.document.body) ? window : t || window;
  }
  function Ee() {
    const e = x.before.value;
    if (!e)
      return 0;
    const t = m();
    return Z(t) === "vertical" ? e.scrollHeight : e.scrollWidth;
  }
  function ue() {
    const e = x.el.value;
    if (!e)
      return { start: 0, end: 0 };
    const t = m(), r = Z(t), i = r === "vertical";
    let s;
    if (ye()) {
      const a = e.getBoundingClientRect(), h = i ? a.height : a.width;
      let g = -(i ? a.top : a.left), S = i ? window.innerHeight : window.innerWidth;
      g < 0 && (S += g, g = 0), g + S > h && (S = h - g), s = {
        start: g,
        end: g + S
      };
    } else i ? s = {
      start: e.scrollTop,
      end: e.scrollTop + e.clientHeight
    } : s = {
      start: de(e.scrollLeft, r, e),
      end: de(e.scrollLeft, r, e) + e.clientWidth
    };
    return s;
  }
  function en() {
    const e = x.el.value;
    if (!e)
      return { start: 0, end: 0 };
    const t = m();
    if (Z(t) === "vertical") {
      const r = de(e.scrollLeft, "horizontal", e);
      return {
        start: r,
        end: r + e.clientWidth
      };
    }
    return {
      start: e.scrollTop,
      end: e.scrollTop + e.clientHeight
    };
  }
  function tn(e, t, r, i) {
    var g, S, T, f;
    const s = Ee(), a = e.start - i - s, h = e.end + i - s;
    if (D > 0) {
      const v = ((g = r[D - 1]) == null ? void 0 : g.accumulator) ?? 0;
      if (a <= v)
        return !0;
    }
    if (D < t - 1) {
      const v = ((S = r[D]) == null ? void 0 : S.accumulator) ?? Number.POSITIVE_INFINITY;
      if (a > v)
        return !0;
    }
    if (G > 1) {
      const v = ((T = r[G - 2]) == null ? void 0 : T.accumulator) ?? 0;
      if (h <= v)
        return !0;
    }
    if (G < t) {
      const v = ((f = r[G - 1]) == null ? void 0 : f.accumulator) ?? Number.POSITIVE_INFINITY;
      if (h > v)
        return !0;
    }
    return !1;
  }
  function _e(e) {
    const t = m(), r = j(t.itemSize);
    if (r !== null)
      return r;
    const i = we.value[e];
    return (i == null ? void 0 : i.size) || Number(t.minItemSize) || 0;
  }
  function nn(e) {
    const t = m();
    return zn(e, {
      direction: Z(t),
      mode: se(t),
      itemSize: j(t.itemSize),
      gridItems: t.gridItems,
      itemSecondarySize: t.itemSecondarySize
    });
  }
  function $(e) {
    var s;
    const t = m(), r = t.gridItems || 1, i = j(t.itemSize);
    return e <= 0 ? 0 : i !== null ? Math.floor(e / r) * i : ((s = we.value[e - 1]) == null ? void 0 : s.accumulator) || 0;
  }
  function yt(e) {
    const t = m(), r = z.value.length, i = t.gridItems || 1, s = j(t.itemSize);
    if (!r)
      return 0;
    if (s !== null) {
      const S = Math.floor(e / s) * i;
      return Math.min(Math.max(S, 0), r - 1);
    }
    let a = 0, h = r - 1, g = 0;
    for (; a <= h; ) {
      const S = Math.floor((a + h) / 2);
      $(S) <= e ? (g = S, a = S + 1) : h = S - 1;
    }
    return g;
  }
  function Q() {
    H && (clearTimeout(H), H = null), J = null;
  }
  function rn() {
    H && clearTimeout(H), H = setTimeout(() => {
      J = null, H = null;
    }, 150);
  }
  function on(e, t) {
    if (!e.length) {
      Q();
      return;
    }
    const r = Math.max(ue().start - Ee(), 0), i = Math.min(yt(r), e.length - 1), s = e[i], a = t ? Oe(s, i, t) : i, h = Ee() + $(i);
    J = {
      key: a,
      offset: ue().start - h
    };
  }
  function wt(e) {
    if (!J)
      return !1;
    const t = m(), r = e ?? z.value, i = q.value ? null : t.keyField, a = Ae(r, i).indexOf(J.key);
    if (a === -1)
      return Q(), !1;
    const h = Ee() + $(a) + J.offset, g = ue().start;
    return Math.abs(h - g) < 0.5 ? !1 : (De = !0, Pe(h), qe(() => {
      De = !1;
    }), !0);
  }
  function Fe() {
    if (xe(), !M.value)
      return;
    const e = ye() ? Kt() : x.el.value;
    e && (pe = e, pe.addEventListener("scroll", Je, Cn() ? { passive: !0 } : !1), ye() && (Ie = e, Ie.addEventListener("resize", Ze)));
  }
  function xe() {
    pe && (pe.removeEventListener("scroll", Je), pe = null), Ie && (Ie.removeEventListener("resize", Ze), Ie = null);
  }
  function ln(e, t, r, i, s, a) {
    const h = Math.ceil(e / t) * r, g = Math.max(0, Math.floor(s.start / r)), S = Math.min(Math.ceil(s.end / r), Math.ceil(e / t)), T = Math.max(0, Math.floor(a.start / i)), f = Math.min(Math.ceil(a.end / i), t), v = [];
    for (let Y = g; Y < S; Y++) {
      const ae = Y * t;
      for (let b = T; b < f; b++) {
        const I = ae + b;
        if (I >= e)
          break;
        v.push(I);
      }
    }
    const C = v[0] ?? 0, k = v.at(-1) ?? -1;
    return {
      renderedIndices: v,
      startIndex: C,
      endIndex: k + 1,
      visibleStartIndex: C,
      visibleEndIndex: k,
      totalSize: h
    };
  }
  function sn() {
    const e = m(), t = j(e.itemSize);
    if (!e.gridItems || t == null)
      return !1;
    const r = x.el.value;
    if (!r)
      return !1;
    const i = e.itemSecondarySize || t, s = Z(e) === "vertical" ? r.clientWidth : r.clientHeight;
    return i * e.gridItems > s;
  }
  function xt(e, t, r, i) {
    if (Math.abs(e - t) !== 1 || t < 0 || t >= r)
      return e;
    const s = e > t ? $(t) + _e(t) : $(t);
    return Math.abs(s - i) <= Ht ? t : e;
  }
  function bt(e, t, r, i, s) {
    if (Math.abs(e - t) !== 1)
      return e;
    const a = s ? t - 1 : t;
    if (a < 0 || a >= r)
      return e;
    const h = e > t ? $(a) + _e(a) : $(a);
    return Math.abs(h - i) <= Ht ? t : e;
  }
  function un(e, t, r) {
    const i = {
      ...e,
      visibleStartIndex: xt(e.visibleStartIndex, lt, t, r.rawViewportStart),
      visibleEndIndex: bt(e.visibleEndIndex, st, t, r.rawViewportEnd, !1),
      startIndex: xt(e.startIndex, D, t, r.renderStart),
      endIndex: bt(e.endIndex, G, t, r.renderEnd, !0)
    };
    i.startIndex > i.visibleStartIndex && (i.startIndex = i.visibleStartIndex);
    const s = Math.min(t, i.visibleEndIndex + 1);
    return i.endIndex < s && (i.endIndex = s), i.endIndex < i.startIndex && (i.endIndex = i.startIndex), i;
  }
  function _(e, t = !1) {
    var Vt, Et, _t, Ft, Pt, Lt, Ct, Nt;
    if (!M.value)
      return { continuous: !0 };
    const r = m(), i = j(r.itemSize), s = r.gridItems || 1, a = r.itemSecondarySize || i || 0, h = ct, g = r.typeField, S = q.value ? null : r.keyField, T = z.value, f = T.length, v = we.value, C = Be, k = R.value;
    let Y = null, ae = null, b, I, A, F, P;
    const dn = i !== null || f === 0 || v[f - 1] != null;
    if (!f || !dn)
      b = I = F = P = A = 0;
    else if (je)
      b = F = 0, I = P = Math.min(r.prerender, T.length), A = 0;
    else {
      const w = ue(), L = en(), W = $e, te = at, V = { ...w }, be = { ...L };
      if (t) {
        let E = w.start - W;
        E < 0 && (E = -E);
        let ne = L.start - te;
        ne < 0 && (ne = -ne);
        const Re = i === null && tn(w, f, v, r.buffer), N = i === null && (E >= h || Re) || i !== null && E >= i, ke = s > 1 && i != null && ne >= a;
        if (!N && !ke)
          return {
            continuous: !0
          };
      }
      $e = w.start, at = L.start;
      const ce = r.buffer;
      V.start -= ce, V.end += ce, be.start -= ce, be.end += ce;
      let O = 0;
      const Rt = x.before.value;
      Rt && (O = Rt.scrollHeight, V.start -= O);
      const kt = x.after.value;
      if (kt) {
        const E = kt.scrollHeight;
        V.end += E;
      }
      const mn = se(r) === "flow" && i === null && !t && ut && Math.abs(w.start - W) <= Yt && Math.abs(L.start - te) <= Yt;
      if (i === null) {
        let E, ne = 0, Re = f - 1, N = ~~(f / 2), ke;
        do
          ke = N, E = ((Vt = v[N]) == null ? void 0 : Vt.accumulator) ?? 0, E < V.start ? ne = N : N < f - 1 && (((Et = v[N + 1]) == null ? void 0 : Et.accumulator) ?? Number.POSITIVE_INFINITY) > V.start && (Re = N), N = ~~((ne + Re) / 2);
        while (N !== ke);
        for (N < 0 && (N = 0), b = N, A = ((_t = v[f - 1]) == null ? void 0 : _t.accumulator) ?? 0, I = N; I < f && (((Ft = v[I]) == null ? void 0 : Ft.accumulator) ?? Number.POSITIVE_INFINITY) < V.end; I++) ;
        for (I === -1 ? I = T.length - 1 : (I++, I > f && (I = f)), F = b; F < f && O + (((Pt = v[F]) == null ? void 0 : Pt.accumulator) ?? Number.POSITIVE_INFINITY) < V.start; F++) ;
        for (P = F; P < f && O + (((Lt = v[P]) == null ? void 0 : Lt.accumulator) ?? Number.POSITIVE_INFINITY) < V.end; P++) ;
        if (mn) {
          const We = un({
            startIndex: b,
            endIndex: I,
            visibleStartIndex: F,
            visibleEndIndex: P
          }, f, {
            rawViewportStart: w.start - O,
            rawViewportEnd: w.end - O,
            renderStart: w.start - ce - O,
            renderEnd: w.end + ce - O
          });
          b = We.startIndex, I = We.endIndex, F = We.visibleStartIndex, P = We.visibleEndIndex;
        }
      } else if (s > 1) {
        const E = ln(
          f,
          s,
          i,
          a,
          V,
          be
        );
        Y = E.renderedIndices, ae = new Set(Y), b = E.startIndex, I = E.endIndex, F = E.visibleStartIndex, P = E.visibleEndIndex, A = E.totalSize;
      } else {
        b = ~~(V.start / i * s);
        const E = b % s;
        b -= E, I = Math.ceil(V.end / i * s), F = Math.max(0, Math.floor((V.start - O) / i * s)), P = Math.floor((V.end - O) / i * s), b < 0 && (b = 0), I > f && (I = f), F < 0 && (F = 0), P > f && (P = f), A = Math.ceil(f / s) * i;
      }
    }
    I - b > Tn.itemsLimit && an(), me.value = A, he.value = 0, ve.value = se(r) === "flow" ? A : 0;
    let p;
    const Qe = b <= G && I >= D, Mt = se(r) === "flow", Ke = Mt && Qe && !e;
    let et = null;
    if (!Qe || e)
      Jt();
    else {
      const w = (L) => {
        const W = k[L];
        if (W && (p = W, p.nr.used)) {
          const te = ae ? ae.has(p.nr.index) : p.nr.index >= b && p.nr.index < I, V = i || v[p.nr.index] && v[p.nr.index].size;
          (!te || !V) && It(p, Ke);
        }
      };
      if (Ke) {
        for (let L = k.length - 1; L >= 0; L--)
          w(L);
        et = qt();
      } else
        for (let L = 0, W = k.length; L < W; L++)
          w(L);
    }
    let ee, Ce, Ne = null, Tt = 0;
    return Bn(b, I, Y, (w) => {
      var V;
      const L = i || v[w] && v[w].size;
      if (!L)
        return;
      ee = T[w];
      const W = S ? Oe(ee, w, S) : w;
      p = C.get(W);
      let te = !1;
      if (p)
        p.item !== ee && (p.item = ee), p.nr.used || console.warn(`Expected existing view's used flag to be true, got ${p.nr.used}`);
      else {
        if (Ce = ee[g], p = Gt(Ce), p) {
          const be = p.nr.index !== w || p.nr.key !== W;
          p.item = ee, p.nr.index = w, p.nr.key = W, p.nr.type !== Ce && console.warn("Reused view's type does not match pool's type"), be && On(p);
        } else
          p = Dt(k, w, ee, W, Ce);
        C.set(W, p), te = !0;
      }
      i === null ? (p.position = ((V = v[w - 1]) == null ? void 0 : V.accumulator) || 0, p.offset = 0) : (p.position = Math.floor(w / s) * i, p.offset = w % s * a), et && te && Zt(p, w, et), Ne ?? (Ne = p.position), Tt += L;
    }), Mt ? (Ne == null ? (he.value = 0, ve.value = A) : (he.value = Ne, ve.value = Math.max(0, A - Ne - Tt)), Ke || ht()) : (he.value = 0, ve.value = 0), D = b, G = I, lt = F, st = P, ut = !0, r.emitUpdate && ((Nt = (Ct = x.callbacks).onUpdate) == null || Nt.call(Ct, b, I, F, P)), se(r) !== "flow" && (X && clearTimeout(X), X = setTimeout(zt, r.updateInterval + 300)), {
      continuous: Qe
    };
  }
  function an() {
    throw ge = setTimeout(() => {
      ge = null, console.warn("It seems the scroller element isn't scrolling, so it tries to render all the items at once.", "Scroller:", x.el.value), console.warn("Make sure the scroller has a fixed height (or width) and 'overflow-y' (or 'overflow-x') set to 'auto' so it can scroll correctly and only render the items visible in the scroll viewport.");
    }), new Error("Rendered items limit reached");
  }
  function cn() {
    if (sn())
      return !1;
    const e = R.value.filter(({ nr: t }) => t.used);
    for (let t = 1; t < e.length; t++)
      if (e[t].nr.index !== e[t - 1].nr.index + 1)
        return !0;
    return !1;
  }
  function zt() {
    if (M.value) {
      if (se() === "flow") {
        ht();
        return;
      }
      R.value.sort((e, t) => e.nr.index - t.nr.index), cn() && (_(!1), X && clearTimeout(X));
    }
  }
  function fn(e, t) {
    if (!M.value)
      return;
    const r = m(), i = Z(r), s = x.el.value;
    if (!s)
      return;
    const a = Math.max(0, Math.min(e, z.value.length - 1)), h = ue().start, g = _n(s, i, r.pageMode), S = $(a), T = _e(a), f = Wt(
      S,
      T,
      h,
      g,
      t == null ? void 0 : t.align,
      (t == null ? void 0 : t.offset) ?? 0
    );
    if (f == null)
      return;
    Pe(f, t);
    const v = j(r.itemSize);
    if (r.gridItems && v != null) {
      const C = x.el.value;
      if (!C)
        return;
      const k = r.gridItems, Y = r.itemSecondarySize || v, b = a % k * Y, I = i === "vertical" ? "horizontal" : "vertical", A = I === "horizontal" ? de(C.scrollLeft, "horizontal", C) : C.scrollTop, F = I === "horizontal" ? C.clientWidth : C.clientHeight, P = Wt(
        b,
        Y,
        A,
        F,
        t == null ? void 0 : t.align,
        (t == null ? void 0 : t.offset) ?? 0
      );
      P != null && nt(C, I, P, t);
    }
  }
  function Pe(e, t) {
    if (!M.value)
      return;
    const r = m(), i = Z(r), s = x.el.value;
    if (s)
      if (ye()) {
        const a = He(s), h = a.getBoundingClientRect(), g = s.getBoundingClientRect(), S = i === "vertical" ? "top" : "left", T = He(s) === document.documentElement || He(s) === document.body ? i === "vertical" ? window.scrollY : window.scrollX : de(
          i === "vertical" ? a.scrollTop : a.scrollLeft,
          i,
          a
        ), f = g[S] - h[S];
        nt(a.tagName === "HTML" ? window : a, i, e + T + f, t);
      } else
        nt(s, i, e, t);
  }
  const K = m(), Le = z.value;
  return Me = Ae(Le, Le.length > 0 && typeof Le[0] != "object" ? null : K.keyField), Se = Ye(Le, K.typeField), K.cache && Xe(K.cache), M.value && K.prerender && (je = !0, _(!1)), K.gridItems && j(K.itemSize) == null && console.error("[vue-recycle-scroller] You must provide an itemSize when using gridItems"), hn(() => {
    M.value && (Fe(), tt(() => {
      je = !1, _(!0), ie.value = !0;
    }));
  }), vn(() => {
    if (!M.value)
      return;
    const e = $e;
    typeof e == "number" && tt(() => {
      Pe(e);
    });
  }), pn(() => {
    gt(), St(), xe();
  }), B(M, (e) => {
    e ? (Fe(), tt(() => {
      _(!0), ie.value = !0;
    })) : (xe(), gt(), St(), ie.value = !1);
  }), B(() => m().cache, (e) => {
    M.value && (Xe(e), _(!0));
  }), B(() => re(m().items).slice(), (e, t) => {
    if (!M.value)
      return;
    const r = m(), i = q.value ? null : r.keyField, s = Ae(e, i), a = Ye(e, r.typeField), h = Me, S = !Yn(
      s,
      a,
      h,
      Se
    );
    r.shift && xn(h, s) > 0 ? on(t ?? [], i) : Q(), Me = s, Se = a, wt(e), _(S);
  }), B(() => m().keyField, () => {
    if (!M.value)
      return;
    const e = m(), t = q.value ? null : e.keyField;
    Me = Ae(z.value, t), Se = Ye(z.value, e.typeField), Q(), _(!0);
  }), B(() => m().typeField, () => {
    if (!M.value)
      return;
    const e = m();
    Se = Ye(z.value, e.typeField), Q(), _(!0);
  }), B(ye, () => {
    M.value && (xe(), Fe(), _(!1));
  }), B(x.el, () => {
    M.value && (xe(), Fe(), _(!1));
  }), B(we, () => {
    M.value && (wt() && rn(), _(!1));
  }), B(() => m().gridItems, () => {
    M.value && _(!0);
  }), B(() => m().itemSecondarySize, () => {
    M.value && _(!0);
  }), {
    pool: R,
    visiblePool: $t,
    totalSize: me,
    startSpacerSize: he,
    endSpacerSize: ve,
    ready: ie,
    sizes: we,
    simpleArray: q,
    scrollToItem: fn,
    scrollToPosition: Pe,
    getScroll: ue,
    findItemIndex: yt,
    getItemOffset: $,
    getItemSize: _e,
    getViewStyle: nn,
    cacheSnapshot: jt,
    restoreCache: Xe,
    updateVisibleItems: _,
    handleResize: Ze,
    handleVisibilityChange: Qt,
    sortViews: zt
  };
}
export {
  Ot as a,
  Ae as b,
  Tn as c,
  wn as d,
  zn as e,
  xn as f,
  j as g,
  bn as r,
  jn as u
};
