// @vitejs/devtools-kit/dist/node/index.js
import { DEFAULT_STATE_USER_SETTINGS } from "../constants.js";
import { REMOTE_CONNECTION_KEY } from "devframe/constants";
import { createEventEmitter } from "devframe/utils/events";
import { nanoid } from "devframe/utils/nanoid";
import { createHostContext, createStorage } from "devframe/node";
import { debounce } from "perfect-debounce";
import { colors } from "devframe/utils/colors";
import { consoleReporter, createLogger, defineDiagnostics } from "logs-sdk";
import { ansiFormatter } from "logs-sdk/formatters/ansi";
import { getInternalContext, resolveBasePath } from "devframe/node/internal";
import { join, resolve } from "pathe";
import process from "node:process";
import { toDataURL } from "mlly";
import { homedir } from "node:os";
import { join as join$1 } from "node:path";
import { serveStaticNodeMiddleware } from "devframe/utils/serve-static";
const logger = createLogger({
	diagnostics: [defineDiagnostics({
		docsBase: "https://devtools.vite.dev/errors",
		codes: {
			DTK0050: {
				message: (p) => `Dock with id "${p.id}" is already registered`,
				hint: "Use the `force` parameter to overwrite an existing registration."
			},
			DTK0051: { message: "Cannot change the id of a dock. Use register() to add new docks." },
			DTK0052: { message: (p) => `Dock with id "${p.id}" is not registered. Use register() to add new docks.` },
			DTK0053: { message: (p) => `Terminal session with id "${p.id}" already registered` },
			DTK0054: { message: (p) => `Terminal session with id "${p.id}" not registered` },
			DTK0055: { message: (p) => `Command "${p.id}" is already registered` },
			DTK0056: { message: "Cannot change the id of a command. Use register() to add new commands." },
			DTK0057: { message: (p) => `Command "${p.id}" is not registered` }
		}
	})],
	formatter: ansiFormatter(colors),
	reporters: consoleReporter
});
//#endregion
//#region src/node/host-commands.ts
var DevToolsCommandsHost = class {
	context;
	commands = /* @__PURE__ */ new Map();
	events = createEventEmitter();
	constructor(context) {
		this.context = context;
	}
	register(command) {
		if (this.commands.has(command.id)) throw logger.DTK0055({ id: command.id }).throw();
		this.commands.set(command.id, command);
		this.events.emit("command:registered", this.toSerializable(command));
		return {
			id: command.id,
			update: (patch) => {
				if ("id" in patch) throw logger.DTK0056().throw();
				const existing = this.commands.get(command.id);
				if (!existing) throw logger.DTK0057({ id: command.id }).throw();
				Object.assign(existing, patch);
				this.events.emit("command:registered", this.toSerializable(existing));
			},
			unregister: () => this.unregister(command.id)
		};
	}
	unregister(id) {
		const deleted = this.commands.delete(id);
		if (deleted) this.events.emit("command:unregistered", id);
		return deleted;
	}
	async execute(id, ...args) {
		const found = this.findCommand(id);
		if (!found) throw logger.DTK0057({ id }).throw();
		if (!found.handler) throw new Error(`Command "${id}" has no handler (group-only command)`);
		return found.handler(...args);
	}
	list() {
		return Array.from(this.commands.values()).map((cmd) => this.toSerializable(cmd));
	}
	findCommand(id) {
		const topLevel = this.commands.get(id);
		if (topLevel) return topLevel;
		for (const cmd of this.commands.values()) if (cmd.children) {
			const child = cmd.children.find((c) => c.id === id);
			if (child) return child;
		}
	}
	toSerializable(cmd) {
		const { handler: _, children, ...rest } = cmd;
		return {
			...rest,
			source: "server",
			...children ? { children: children.map((c) => this.toSerializable(c)) } : {}
		};
	}
};
//#endregion
//#region src/node/host-docks.ts
function normaliseRemoteOptions(remote) {
	const opts = remote === true ? {} : remote;
	return {
		transport: opts.transport ?? "fragment",
		originLock: opts.originLock ?? true
	};
}
function base64UrlEncode(value) {
	const bytes = new TextEncoder().encode(value);
	let binary = "";
	for (const byte of bytes) binary += String.fromCharCode(byte);
	return btoa(binary).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}
function buildRemoteUrl(baseUrl, payload, transport) {
	const param = `${REMOTE_CONNECTION_KEY}=${base64UrlEncode(JSON.stringify(payload))}`;
	if (transport === "fragment") {
		const hashIdx = baseUrl.indexOf("#");
		if (hashIdx === -1) return `${baseUrl}#${param}`;
		return `${baseUrl.slice(0, hashIdx)}#${param}`;
	}
	const qIdx = baseUrl.indexOf("?");
	const hashIdx = baseUrl.indexOf("#");
	const hash = hashIdx === -1 ? "" : baseUrl.slice(hashIdx);
	const beforeHash = hashIdx === -1 ? baseUrl : baseUrl.slice(0, hashIdx);
	return `${beforeHash}${qIdx === -1 || qIdx >= (hashIdx === -1 ? beforeHash.length : hashIdx) ? "?" : "&"}${param}${hash}`;
}
var DevToolsDockHost = class {
	context;
	views = /* @__PURE__ */ new Map();
	events = createEventEmitter();
	userSettings = void 0;
	/** Dock-id → allocated remote token + resolved options. */
	remoteDocks = /* @__PURE__ */ new Map();
	constructor(context) {
		this.context = context;
	}
	async init() {
		this.userSettings = await this.context.rpc.sharedState.get("devframe:user-settings", { sharedState: createStorage({
			filepath: join(this.context.host.getStorageDir("workspace"), "settings.json"),
			initialValue: DEFAULT_STATE_USER_SETTINGS()
		}) });
	}
	values({ includeBuiltin = true } = {}) {
		const context = this.context;
		const builtinDocksEntries = [
			{
				type: "~builtin",
				id: "~terminals",
				title: "Terminals",
				icon: "ph:terminal-duotone",
				category: "~builtin",
				get when() {
					return context.terminals.sessions.size === 0 ? "false" : void 0;
				}
			},
			{
				type: "~builtin",
				id: "~messages",
				title: "Messages & Notifications",
				icon: "ph:notification-duotone",
				category: "~builtin",
				get badge() {
					const size = context.messages.entries.size;
					return size > 0 ? String(size) : void 0;
				}
			},
			{
				type: "~builtin",
				id: "~settings",
				title: "Settings",
				category: "~builtin",
				icon: "ph:gear-duotone"
			}
		];
		return [...Array.from(this.views.values(), (view) => this.projectView(view)), ...includeBuiltin ? builtinDocksEntries : []];
	}
	projectView(view) {
		if (view.type !== "iframe" || !view.remote) return view;
		const record = this.remoteDocks.get(view.id);
		const endpoint = getInternalContext(this.context).wsEndpoint;
		if (!record || !endpoint) return view;
		const payload = {
			v: 1,
			backend: "websocket",
			websocket: endpoint.url,
			authToken: record.token,
			origin: this.resolveDevServerOrigin()
		};
		return {
			...view,
			url: buildRemoteUrl(view.url, payload, record.options.transport)
		};
	}
	resolveDevServerOrigin() {
		return this.context.host.resolveOrigin();
	}
	register(view, force) {
		if (this.views.has(view.id) && !force) throw logger.DTK0050({ id: view.id }).throw();
		this.prepareRemoteRegistration(view);
		this.views.set(view.id, view);
		this.events.emit("dock:entry:updated", view);
		return { update: (patch) => {
			if (patch.id && patch.id !== view.id) throw logger.DTK0051().throw();
			this.update(Object.assign(this.views.get(view.id), patch));
		} };
	}
	update(view) {
		if (!this.views.has(view.id)) throw logger.DTK0052({ id: view.id }).throw();
		this.prepareRemoteRegistration(view);
		this.views.set(view.id, view);
		this.events.emit("dock:entry:updated", view);
	}
	prepareRemoteRegistration(view) {
		const internal = getInternalContext(this.context);
		internal.revokeRemoteTokensForDock(view.id);
		this.remoteDocks.delete(view.id);
		if (view.type !== "iframe" || !view.remote) return;
		const options = normaliseRemoteOptions(view.remote);
		let dockOrigin;
		try {
			dockOrigin = new URL(view.url).origin;
		} catch {
			dockOrigin = this.resolveDevServerOrigin();
		}
		const token = internal.allocateRemoteToken(view.id, dockOrigin, options.originLock);
		this.remoteDocks.set(view.id, {
			token,
			options
		});
	}
};
//#endregion
//#region src/node/host-messages.ts
const MAX_ENTRIES = 1e3;
var DevToolsMessagesHost = class {
	context;
	entries = /* @__PURE__ */ new Map();
	events = createEventEmitter();
	/** Tracks when each entry was last added or updated (monotonic) */
	lastModified = /* @__PURE__ */ new Map();
	/** Tracks recently removed entry IDs with their removal time */
	removals = [];
	_autoDeleteTimers = /* @__PURE__ */ new Map();
	_clock = 0;
	_tick() {
		return ++this._clock;
	}
	constructor(context) {
		this.context = context;
	}
	async add(input) {
		if (input.id && this.entries.has(input.id)) {
			await this.update(input.id, input);
			return this._createHandle(input.id);
		}
		const entry = {
			...input,
			id: input.id ?? nanoid(),
			timestamp: input.timestamp ?? Date.now(),
			from: input.from ?? "server"
		};
		if (this.entries.size >= MAX_ENTRIES) {
			const oldest = this.entries.keys().next().value;
			await this.remove(oldest);
		}
		this.entries.set(entry.id, entry);
		this.lastModified.set(entry.id, this._tick());
		this.events.emit("message:added", entry);
		if (entry.autoDelete) this._autoDeleteTimers.set(entry.id, setTimeout(() => {
			this.remove(entry.id);
		}, entry.autoDelete));
		return this._createHandle(entry.id);
	}
	async update(id, patch) {
		const existing = this.entries.get(id);
		if (!existing) return void 0;
		const updated = {
			...existing,
			...patch,
			id: existing.id,
			from: existing.from,
			timestamp: existing.timestamp
		};
		this.entries.set(id, updated);
		this.lastModified.set(id, this._tick());
		this.events.emit("message:updated", updated);
		if (patch.autoDelete !== void 0) {
			const timer = this._autoDeleteTimers.get(id);
			if (timer) {
				clearTimeout(timer);
				this._autoDeleteTimers.delete(id);
			}
			if (patch.autoDelete) this._autoDeleteTimers.set(id, setTimeout(() => {
				this.remove(id);
			}, patch.autoDelete));
		}
		return updated;
	}
	async remove(id) {
		const timer = this._autoDeleteTimers.get(id);
		if (timer) {
			clearTimeout(timer);
			this._autoDeleteTimers.delete(id);
		}
		this.entries.delete(id);
		this.lastModified.delete(id);
		this.removals.push({
			id,
			time: this._tick()
		});
		this.events.emit("message:removed", id);
	}
	async clear() {
		for (const timer of this._autoDeleteTimers.values()) clearTimeout(timer);
		this._autoDeleteTimers.clear();
		const tick = this._tick();
		for (const id of this.entries.keys()) this.removals.push({
			id,
			time: tick
		});
		this.entries.clear();
		this.lastModified.clear();
		this.events.emit("message:cleared");
	}
	_createHandle(id) {
		const host = this;
		return {
			get entry() {
				return host.entries.get(id);
			},
			get id() {
				return id;
			},
			update: (patch) => host.update(id, patch),
			dismiss: () => host.remove(id)
		};
	}
};
//#endregion
//#region src/node/host-terminals.ts
/**
* Channel name used for terminal stream output. Stable, well-known so the
* standalone client (`packages/core/src/client/webcomponents/state/terminals.ts`)
* can subscribe by name.
*/
const TERMINAL_STREAM_CHANNEL = "devframe:terminals";
const TERMINAL_REPLAY_WINDOW = 1e3;
var DevToolsTerminalHost = class {
	context;
	sessions = /* @__PURE__ */ new Map();
	events = createEventEmitter();
	_boundStreams = /* @__PURE__ */ new Map();
	_channel;
	constructor(context) {
		this.context = context;
	}
	/**
	* Lazily acquire the streaming channel — `context.rpc` isn't assigned
	* until after every host is constructed, so we can't grab it in the
	* constructor.
	*/
	getStreamingChannel() {
		if (this._channel) return this._channel;
		if (!this.context.rpc?.streaming) return void 0;
		this._channel = this.context.rpc.streaming.create(TERMINAL_STREAM_CHANNEL, { replayWindow: TERMINAL_REPLAY_WINDOW });
		return this._channel;
	}
	register(session) {
		if (this.sessions.has(session.id)) throw logger.DTK0053({ id: session.id }).throw();
		this.sessions.set(session.id, session);
		this.bindStream(session);
		this.events.emit("terminal:session:updated", session);
		return session;
	}
	update(patch) {
		if (!this.sessions.has(patch.id)) throw logger.DTK0054({ id: patch.id }).throw();
		const session = this.sessions.get(patch.id);
		Object.assign(session, patch);
		this.sessions.set(patch.id, session);
		this.bindStream(session);
		this.events.emit("terminal:session:updated", session);
	}
	remove(session) {
		this._boundStreams.get(session.id)?.dispose();
		this.sessions.delete(session.id);
		this.events.emit("terminal:session:updated", session);
		this._boundStreams.delete(session.id);
	}
	bindStream(session) {
		if (this._boundStreams.has(session.id) && this._boundStreams.get(session.id)?.stream === session.stream) return;
		this._boundStreams.get(session.id)?.dispose();
		this._boundStreams.delete(session.id);
		if (!session.stream) return;
		session.buffer ||= [];
		const sessionBuffer = session.buffer;
		const sink = this.getStreamingChannel()?.start({ id: session.id });
		const writer = new WritableStream({
			write(chunk) {
				sessionBuffer.push(chunk);
				sink?.write(chunk);
			},
			close() {
				sink?.close();
			},
			abort(reason) {
				sink?.error(reason);
			}
		});
		session.stream.pipeTo(writer).catch(() => {});
		this._boundStreams.set(session.id, {
			dispose: () => {
				if (sink && !sink.closed) sink.close();
			},
			stream: session.stream
		});
	}
	async startChildProcess(executeOptions, terminal) {
		if (this.sessions.has(terminal.id)) throw logger.DTK0053({ id: terminal.id }).throw();
		const { exec } = await import("tinyexec");
		let controller;
		const stream = new ReadableStream({ start(_controller) {
			controller = _controller;
		} });
		function createChildProcess() {
			const cp = exec(executeOptions.command, executeOptions.args || [], { nodeOptions: {
				env: {
					COLORS: "true",
					FORCE_COLOR: "true",
					...executeOptions.env || {}
				},
				cwd: executeOptions.cwd ?? process.cwd(),
				stdio: "pipe"
			} });
			(async () => {
				for await (const chunk of cp) controller?.enqueue(chunk);
			})();
			return cp;
		}
		let cp = createChildProcess();
		const restart = async () => {
			cp?.kill();
			cp = createChildProcess();
		};
		const terminate = async () => {
			cp?.kill();
			cp = void 0;
		};
		const session = {
			...terminal,
			status: "running",
			stream,
			type: "child-process",
			executeOptions,
			getChildProcess: () => cp?.process,
			terminate,
			restart
		};
		this.register(session);
		return Promise.resolve(session);
	}
};
//#endregion
//#region src/node/context.ts
/**
* Create a kit-level node context: wraps devframe's `createHostContext`,
* attaches the hub hosts (`docks`, `terminals`, `messages`, `commands`),
* and wires the shared-state synchronization that powers the unified
* client UI.
*/
async function createKitContext(options) {
	const context = await createHostContext(options);
	const docks = new DevToolsDockHost(context);
	const terminals = new DevToolsTerminalHost(context);
	const messages = new DevToolsMessagesHost(context);
	const commands = new DevToolsCommandsHost(context);
	context.docks = docks;
	context.terminals = terminals;
	context.messages = messages;
	context.commands = commands;
	if (options.viteConfig) Object.defineProperty(context, "viteConfig", {
		value: options.viteConfig,
		enumerable: true
	});
	if (options.viteServer) Object.defineProperty(context, "viteServer", {
		value: options.viteServer,
		enumerable: true
	});
	await docks.init();
	let jrCounter = 0;
	context.createJsonRenderer = (initialSpec) => {
		const stateKey = `devframe:json-render:${jrCounter++}`;
		const statePromise = context.rpc.sharedState.get(stateKey, { initialValue: initialSpec });
		return {
			_stateKey: stateKey,
			async updateSpec(spec) {
				(await statePromise).mutate(() => spec);
			},
			async updateState(newState) {
				(await statePromise).mutate((draft) => {
					draft.state = {
						...draft.state,
						...newState
					};
				});
			}
		};
	};
	const debounceMs = options.mode === "build" ? 0 : 10;
	const docksSharedState = await context.rpc.sharedState.get("devframe:docks", { initialValue: [] });
	const refreshDocks = debounce(() => {
		docksSharedState.mutate(() => docks.values());
	}, debounceMs);
	docks.events.on("dock:entry:updated", refreshDocks);
	const broadcastTerminals = debounce(() => {
		context.rpc.broadcast({
			method: "devframe:terminals:updated",
			args: []
		});
		docksSharedState.mutate(() => docks.values());
	}, debounceMs);
	terminals.events.on("terminal:session:updated", broadcastTerminals);
	const broadcastMessages = debounce(() => {
		context.rpc.broadcast({
			method: "devframe:messages:updated",
			args: []
		});
		docksSharedState.mutate(() => docks.values());
	}, debounceMs);
	messages.events.on("message:added", broadcastMessages);
	messages.events.on("message:updated", broadcastMessages);
	messages.events.on("message:removed", broadcastMessages);
	messages.events.on("message:cleared", broadcastMessages);
	const commandsSharedState = await context.rpc.sharedState.get("devframe:commands", { initialValue: [] });
	const syncCommands = debounce(() => {
		commandsSharedState.mutate(() => commands.list());
	}, debounceMs);
	commands.events.on("command:registered", syncCommands);
	commands.events.on("command:unregistered", syncCommands);
	return context;
}
//#endregion
//#region src/node/create-plugin-from-devframe.ts
/**
* Wrap a {@link DevframeDefinition} as a Vite plugin that mounts inside
* `@vitejs/devtools` (Vite DevTools). The kit takes care of mounting
* the SPA at the resolved base path, synthesizing an iframe dock entry
* from the definition's metadata, and threading the kit-augmented
* context into both the devframe-level `d.setup` and the optional
* `options.setup` hook.
*
* For richer kit-specific behavior (registering terminals/commands,
* adding additional dock entries), use `options.setup`.
*/
function createPluginFromDevframe(d, options = {}) {
	const base = options.base ?? resolveBasePath(d, "hosted");
	return {
		name: options.name ?? `devframe:${d.id}`,
		devtools: {
			capabilities: options.capabilities ?? d.capabilities,
			async setup(rawCtx) {
				const ctx = rawCtx;
				if (d.cli?.distDir) ctx.views.hostStatic(base, resolve(d.cli.distDir));
				ctx.docks.register({
					id: d.id,
					title: d.name,
					icon: d.icon ?? "ph:plug-duotone",
					...options.dock,
					type: "iframe",
					url: base
				});
				await d.setup(ctx);
				if (options.setup) await options.setup(ctx);
			}
		}
	};
}
//#endregion
//#region src/node/utils.ts
/**
* Create a quick `ClientScriptEntry` from an inline function or
* stringified code. Useful for prototyping `action` / `renderer`
* dock entries without setting up a separate importable module.
*
* @experimental Prefer a proper importable module for production use.
*/
function createSimpleClientScript(fn) {
	return {
		importFrom: toDataURL(`const fn = ${fn.toString()}; export default fn`),
		importName: "default"
	};
}
//#endregion
//#region src/node/vite-host.ts
function createViteDevToolsHost(options) {
	const { viteConfig, viteServer } = options;
	const workspaceRoot = options.workspaceRoot ?? viteConfig.root;
	return {
		mountStatic(base, distDir) {
			if (viteConfig.command !== "serve") return;
			if (!viteServer) throw new Error("viteServer is required to mount static assets in dev mode");
			viteServer.middlewares.use(base, serveStaticNodeMiddleware(distDir));
		},
		resolveOrigin() {
			const resolved = viteServer?.resolvedUrls?.local?.[0];
			if (resolved) return new URL(resolved).origin;
			const https = !!viteConfig.server.https;
			const host = typeof viteConfig.server.host === "string" ? viteConfig.server.host : "localhost";
			const port = viteConfig.server.port ?? (https ? 443 : 80);
			return `${https ? "https" : "http"}://${host === "0.0.0.0" || host === "::" || !host ? "localhost" : host}:${port}`;
		},
		getStorageDir(scope) {
			return scope === "workspace" ? join$1(workspaceRoot, "node_modules/.vite/devtools") : join$1(homedir(), ".vite/devtools");
		}
	};
}
//#endregion
export { DevToolsCommandsHost, DevToolsDockHost, DevToolsMessagesHost, DevToolsTerminalHost, createKitContext, createPluginFromDevframe, createSimpleClientScript, createViteDevToolsHost };
