// devframe/dist/constants.mjs
//#region src/constants.ts
const DEVTOOLS_MOUNT_PATH = "/__devtools/";
const DEVTOOLS_MOUNT_PATH_NO_TRAILING_SLASH = "/__devtools";
const DEVTOOLS_DIRNAME = "__devtools";
const DEVTOOLS_CONNECTION_META_FILENAME = "__connection.json";
const DEVTOOLS_RPC_DUMP_MANIFEST_FILENAME = "__rpc-dump/index.json";
const DEVTOOLS_DOCK_IMPORTS_FILENAME = "__client-imports.js";
const DEVTOOLS_DOCK_IMPORTS_VIRTUAL_ID = "/__devtools-client-imports.js";
const DEVTOOLS_RPC_DUMP_DIRNAME = "__rpc-dump";
/**
* URL fragment / query parameter name carrying the remote dock
* connection descriptor (defined as `RemoteConnectionInfo` in
* `@vitejs/devtools-kit`) injected into remote-UI iframe dock URLs.
*/
const REMOTE_CONNECTION_KEY = "vite-devtools-kit-connection";
//#endregion
export { DEVTOOLS_CONNECTION_META_FILENAME, DEVTOOLS_DIRNAME, DEVTOOLS_DOCK_IMPORTS_FILENAME, DEVTOOLS_DOCK_IMPORTS_VIRTUAL_ID, DEVTOOLS_MOUNT_PATH, DEVTOOLS_MOUNT_PATH_NO_TRAILING_SLASH, DEVTOOLS_RPC_DUMP_DIRNAME, DEVTOOLS_RPC_DUMP_MANIFEST_FILENAME, REMOTE_CONNECTION_KEY };
