// vue-virtual-scroller/dist/composables/useDynamicScroller.js
import { nextTick as ie, reactive as ve, computed as K, toValue as q, provide as j, shallowRef as G, watchEffect as pe, shallowReactive as ne, triggerRef as ge, watch as M, onActivated as Ne, onDeactivated as Qe, onUnmounted as qe, effectScope as je } from "vue";
import { a as Ge, b as V, d as he, u as Ye, f as Je, e as Re, r as re } from "../useRecycleScroller-G2SVHG1u.js";
import { r as Ve, a as Ie } from "../keyField-D6okV6nC.js";
import { c as Xe } from "../dynamicScrollerMeasurement-Cq5NQLJa.js";
function Ze() {
  const n = /* @__PURE__ */ new Map();
  return {
    on(i, o) {
      let a = n.get(i);
      a || (a = /* @__PURE__ */ new Set(), n.set(i, a)), a.add(o);
    },
    off(i, o) {
      const a = n.get(i);
      a && (a.delete(o), a.size || n.delete(i));
    },
    emit(i, o) {
      const a = n.get(i);
      if (a != null && a.size)
        for (const d of [...a])
          d(o);
    },
    clear() {
      n.clear();
    }
  };
}
function et(n = {}) {
  const i = n.queueFlush ?? ((m) => ie(m)), o = n.overflowQueueFlush ?? i, a = Math.max(1, n.maxTasksPerFlush ?? Number.POSITIVE_INFINITY);
  let d = !1;
  const p = /* @__PURE__ */ new Map();
  function R() {
    d || !p.size || (d = !0, i(A));
  }
  function A() {
    if (d = !1, !p.size)
      return;
    const m = [];
    for (const [h, D] of p)
      if (m.push(D), p.delete(h), m.length >= a)
        break;
    const I = m.map((h) => ({
      task: h,
      value: h.read()
    }));
    for (const h of I)
      h.value != null && h.task.write(h.value), h.task.done();
    p.size && (o(A), d = !0);
  }
  function x(m, I) {
    p.set(m, I), R();
  }
  function y(m) {
    p.delete(m);
  }
  return {
    schedule: x,
    cancel: y
  };
}
function tt(n) {
  for (let i = 0; i < n.length; i++)
    n[i];
}
function ke(n) {
  return n.item;
}
function L(n, i) {
  const o = i.get(n);
  return o || n;
}
function nt(n) {
  const i = {};
  return Object.defineProperties(i, {
    item: {
      enumerable: !0,
      get: () => n.item.item
    },
    itemWithSize: {
      enumerable: !0,
      get: () => n.item
    },
    size: {
      enumerable: !0,
      get: () => n.item.size
    },
    position: {
      enumerable: !0,
      get: () => n.position
    },
    offset: {
      enumerable: !0,
      get: () => n.offset
    },
    id: {
      enumerable: !0,
      get: () => n.nr.id
    },
    index: {
      enumerable: !0,
      get: () => n.nr.index
    },
    used: {
      enumerable: !0,
      get: () => n.nr.used
    },
    key: {
      enumerable: !0,
      get: () => n.nr.key
    },
    type: {
      enumerable: !0,
      get: () => n.nr.type
    }
  }), i;
}
function it(n) {
  return n._vs_styleStamp ?? 0;
}
function rt(n) {
  return n._vs_visibilityStamp ?? 0;
}
const Fe = [
  "position",
  "top",
  "left",
  "transform",
  "willChange",
  "visibility",
  "pointerEvents",
  "display"
];
function ye(n) {
  const i = {};
  for (const o of Fe)
    i[o] = n.style[o];
  return i;
}
function Te(n, i) {
  for (const o of Fe)
    n.style[o] = i[o] ?? "";
}
function B(n, i, o) {
  n.style[i] !== o && (n.style[i] = o);
}
function Se(n, i, o, a, d, p, R) {
  if (!("view" in i)) {
    Te(n, p);
    return;
  }
  const A = n.tagName === "TR", x = d ? re({
    direction: o,
    flowMode: !0
  }) : re({
    direction: o,
    disableTransform: a || A
  }), y = Re(L(i.view, R), {
    direction: o,
    mode: x
  });
  B(n, "position", String(y.position ?? "")), B(n, "top", String(y.top ?? "")), B(n, "left", String(y.left ?? "")), B(n, "transform", x === "flow" || A ? "" : String(y.transform ?? "")), B(n, "willChange", String(y.willChange ?? "")), B(n, "visibility", String(y.visibility ?? "")), B(n, "pointerEvents", String(y.pointerEvents ?? "")), B(n, "display", String(y.display ?? ""));
}
function ze(n, i) {
  return "view" in n && !i;
}
function be(n, i, o, a, d, p) {
  if ("view" in n) {
    const R = L(n.view, p);
    return {
      active: R.nr.used && d,
      id: ke(R).id
    };
  }
  return {
    active: i.active && d,
    id: ct(
      i.item,
      o,
      a,
      i.index
    )
  };
}
function st(n, i, o, a) {
  if (!("view" in n))
    return {
      direction: i,
      flowMode: o,
      legacy: !0
    };
  if (o) {
    const p = L(n.view, a);
    return {
      direction: i,
      flowMode: o,
      legacy: !1,
      visibilityStamp: rt(p)
    };
  }
  const d = L(n.view, a);
  return {
    active: d.nr.used,
    direction: i,
    flowMode: o,
    legacy: !1,
    styleStamp: it(d),
    position: ze(n, o) ? d.position : null,
    offset: ze(n, o) ? d.offset : null
  };
}
function we(n, i) {
  if ("view" in n) {
    const o = L(n.view, i);
    return {
      item: ke(o).item,
      active: o.nr.used,
      index: o.nr.index,
      watchData: n.watchData ?? !1,
      emitResize: n.emitResize ?? !1,
      sizeDependencies: n.sizeDependencies ?? null,
      onResize: n.onResize
    };
  }
  return {
    watchData: !1,
    emitResize: !1,
    sizeDependencies: null,
    ...n
  };
}
function ot(n, i) {
  n.item !== i.item && (n.item = i.item), n.active !== i.active && (n.active = i.active), n.index !== i.index && (n.index = i.index), n.watchData !== i.watchData && (n.watchData = i.watchData), n.sizeDependencies !== i.sizeDependencies && (n.sizeDependencies = i.sizeDependencies), n.emitResize !== i.emitResize && (n.emitResize = i.emitResize), n.onResize !== i.onResize && (n.onResize = i.onResize);
}
function lt(n, i) {
  n.onResize !== i.onResize && (n.onResize = i.onResize);
}
function ct(n, i, o, a) {
  return o ? a ?? null : Ie(n, a, i);
}
function mt(n) {
  const i = Ge(n);
  let o = 0, a = {};
  const d = Ze();
  let p = !1, R, A = !1, x = [], y = null, m = null, I = null;
  const h = /* @__PURE__ */ new Set(), D = [], v = ve({
    active: !0,
    sizes: {},
    keyField: c().keyField,
    simpleArray: !1
  }), P = K(() => {
    const e = q(c().items);
    return tt(e), e;
  }), C = K(() => c().direction ?? "vertical"), b = K(() => q(c().el)), _e = K(() => q(c().before)), Me = K(() => q(c().after)), $ = /* @__PURE__ */ new Map();
  function c() {
    return i.value;
  }
  const z = K(() => q(c().enabled ?? !0));
  function W(e) {
    let t = -1;
    return t = requestAnimationFrame(() => {
      h.delete(t), e();
    }), h.add(t), t;
  }
  function se() {
    for (const e of h)
      cancelAnimationFrame(e);
    h.clear();
  }
  const Ae = et({
    // Spread row measurement across frames so one huge flush cannot monopolize layout time.
    maxTasksPerFlush: 6,
    overflowQueueFlush(e) {
      W(e);
    }
  });
  typeof ResizeObserver < "u" && (R = new ResizeObserver((e) => {
    W(() => {
      if (Array.isArray(e)) {
        for (const t of e)
          if (t.target && t.target.$_vs_onResize) {
            let r, s;
            if (t.borderBoxSize) {
              const u = t.borderBoxSize[0];
              r = u.inlineSize, s = u.blockSize;
            } else
              r = t.contentRect.width, s = t.contentRect.height;
            t.target.$_vs_onResize(t.target.$_vs_id, r, s);
          }
      }
    });
  }));
  const X = {
    vscrollData: v,
    resizeObserver: R,
    measureQueue: Ae,
    direction: C,
    undefinedMap: a,
    undefinedSizeCount: {
      get value() {
        return o;
      },
      set value(e) {
        o = e;
      }
    },
    onVscrollUpdate(e) {
      const t = (r) => {
        e(r);
      };
      return d.on("vscroll:update", t), () => d.off("vscroll:update", t);
    }
  };
  j("vscrollData", v), j("vscrollParent", {
    get $_undefinedSizes() {
      return o;
    },
    set $_undefinedSizes(e) {
      o = e;
    },
    get $_undefinedMap() {
      return a;
    },
    set $_undefinedMap(e) {
      a = e;
    },
    $_events: d,
    direction: C
  }), j("vscrollResizeObserver", R), j("vscrollMeasurementContext", X), j("vscrollAnchorRegistry", {
    delete(e) {
      $.delete(e);
    },
    set(e, t) {
      $.set(e, t);
    }
  });
  const F = K(() => {
    const e = P.value;
    return e.length > 0 && typeof e[0] != "object";
  }), oe = G(0), N = G(D);
  pe(() => {
    if (!z.value)
      return;
    const e = m || !b.value ? null : xe(b.value.scrollTop, D), r = c().keyField, s = F.value, u = v.sizes, l = P.value, f = l.length;
    let S = D.length !== f;
    for (let g = 0; g < f; g++) {
      const U = l[g], k = s ? g : Ve(U, g, r);
      let E = u[k];
      typeof E > "u" && !a[k] && (E = 0);
      let w = D[g];
      if (!w) {
        w = ne({
          item: U,
          id: k,
          size: E
        }), D[g] = w, S = !0;
        continue;
      }
      w.item !== U && (w.item = U, S = !0), w.id !== k && (w.id = k, S = !0), w.size !== E && (w.size = E, S = !0);
    }
    D.length !== f && (D.length = f, S = !0), S && (y = e, oe.value++, ge(N));
  });
  function xe(e, t) {
    if (!t.length)
      return null;
    const r = c().minItemSize;
    let s = 0, u = 0, l = t[0].id;
    for (const f of t) {
      const S = f.size || r, g = s + S;
      if (u = s, l = f.id, e < g)
        return {
          key: f.id,
          offset: e - s
        };
      s = g;
    }
    return {
      key: l,
      offset: Math.max(0, e - u)
    };
  }
  const Y = c();
  x = V(P.value, F.value ? null : Y.keyField), Y.cache && (v.sizes = he(Y.cache, P.value, F.value ? null : Y.keyField));
  function De() {
    var e, t;
    (t = (e = c()).onHidden) == null || t.call(e);
  }
  function Oe(e, t, r, s) {
    var u, l;
    (l = (u = c()).onUpdate) == null || l.call(u, e, t, r, s);
  }
  const Pe = ve({
    get enabled() {
      return z.value;
    },
    get items() {
      return N.value;
    },
    get keyField() {
      return "id";
    },
    get direction() {
      return C.value;
    },
    get itemSize() {
      return null;
    },
    get gridItems() {
    },
    get itemSecondarySize() {
    },
    get minItemSize() {
      return c().minItemSize;
    },
    get sizeField() {
      return "size";
    },
    get typeField() {
      return "type";
    },
    get buffer() {
      return c().buffer ?? 200;
    },
    get pageMode() {
      return c().pageMode ?? !1;
    },
    get shift() {
      return !1;
    },
    get cache() {
      return c().cache;
    },
    get prerender() {
      return c().prerender ?? 0;
    },
    get emitUpdate() {
      return c().emitUpdate ?? !1;
    },
    get disableTransform() {
      return c().disableTransform ?? !1;
    },
    get flowMode() {
      return c().flowMode ?? !1;
    },
    get hiddenPosition() {
      return c().hiddenPosition;
    },
    get updateInterval() {
      return c().updateInterval ?? 0;
    },
    get el() {
      return b.value;
    },
    get before() {
      return _e.value;
    },
    get after() {
      return Me.value;
    },
    get onResize() {
      return le;
    },
    get onVisible() {
      return ce;
    },
    get onHidden() {
      return De;
    },
    get onUpdate() {
      return Oe;
    }
  });
  function le() {
    var e, t;
    H(), (t = (e = c()).onResize) == null || t.call(e);
  }
  function ce() {
    var e, t;
    d.emit("vscroll:update", { force: !1 }), (t = (e = c()).onVisible) == null || t.call(e);
  }
  const T = Ye(
    Pe
  ), O = /* @__PURE__ */ new WeakMap(), ue = /* @__PURE__ */ new WeakMap(), Z = G([]);
  function Ce(e) {
    const t = c();
    return Re(L(e, O), {
      direction: C.value,
      mode: re({
        direction: C.value,
        disableTransform: t.disableTransform ?? !1,
        flowMode: t.flowMode ?? !1
      })
    });
  }
  function ae(e) {
    let t = ue.get(e);
    return t || (t = nt(e), ue.set(e, t), O.set(t, e)), t;
  }
  pe(() => {
    const e = T.pool.value, t = e.length, r = Z.value;
    let s = r.length !== t;
    for (let u = 0; u < t; u++) {
      const l = ae(e[u]);
      r[u] !== l && (r[u] = l, s = !0);
    }
    r.length !== t && (r.length = t, s = !0), s && ge(Z);
  });
  const Ee = K(
    () => T.visiblePool.value.map(ae)
  ), J = /* @__PURE__ */ new WeakMap();
  function ee() {
    I != null && (cancelAnimationFrame(I), h.delete(I), I = null);
  }
  function _() {
    ee(), m = null;
  }
  function Ke() {
    m == null || I != null || (I = W(() => {
      I = null, te();
    }));
  }
  function Be() {
    const e = b.value;
    if (!e)
      return null;
    const t = e.getBoundingClientRect();
    let r = null;
    for (const [s, u] of $.entries()) {
      if (!u.active || getComputedStyle(s).visibility === "hidden")
        continue;
      const l = s.getBoundingClientRect();
      if (l.bottom <= t.top || l.top >= t.bottom)
        continue;
      const f = Math.max(l.top, t.top) - t.top;
      (!r || f < r.score) && (r = {
        key: u.id,
        offset: l.top - t.top,
        score: f
      });
    }
    return r;
  }
  function We(e) {
    const t = b.value;
    if (!t) {
      _();
      return;
    }
    const r = t.scrollTop, s = Math.min(T.findItemIndex(r), e.length - 1), u = e[s];
    if (u == null) {
      _();
      return;
    }
    const l = Be();
    m = {
      logicalKey: u,
      logicalOffset: r - T.getItemOffset(s),
      pendingKeys: /* @__PURE__ */ new Set(),
      stableFrames: 0,
      visualKey: (l == null ? void 0 : l.key) ?? null,
      visualOffset: (l == null ? void 0 : l.offset) ?? 0
    };
  }
  function fe(e) {
    const t = b.value;
    return !t || Math.abs(t.scrollTop - e) < 0.5 ? !1 : (A = !0, t.scrollTop = e, T.updateVisibleItems(!0), t.dispatchEvent(new Event("scroll")), W(() => {
      A = !1;
    }), !0);
  }
  function te() {
    const e = m, t = b.value;
    if (!e || !t)
      return;
    const r = N.value.findIndex((f) => f.id === e.logicalKey);
    if (r === -1) {
      _();
      return;
    }
    let s = !1;
    const u = T.getItemOffset(r) + e.logicalOffset;
    if (s = fe(u) || s, !s && e.visualKey != null)
      for (const [f, S] of $.entries()) {
        if (!S.active || S.id !== e.visualKey || getComputedStyle(f).visibility === "hidden")
          continue;
        const g = f.getBoundingClientRect().top - t.getBoundingClientRect().top - e.visualOffset;
        s = fe(t.scrollTop + g) || s;
        break;
      }
    let l = !0;
    for (const f of e.pendingKeys)
      if (!(typeof v.sizes[f] == "number" && v.sizes[f] > 0)) {
        l = !1;
        break;
      }
    if (!s && l) {
      if (e.stableFrames++, e.stableFrames >= 2) {
        _();
        return;
      }
    } else
      e.stableFrames = 0;
    Ke();
  }
  function de(e, t, r, s) {
    const u = je(), l = G(t), f = ne({
      ...r
    }), S = ne({
      onResize: r.onResize
    }), g = G(e), U = u.run(() => (M(() => be(
      l.value,
      f,
      c().keyField,
      v.simpleArray,
      v.active,
      O
    ), () => {
      const k = g.value;
      if (k) {
        const E = be(
          l.value,
          f,
          c().keyField,
          v.simpleArray,
          v.active,
          O
        ), w = E.id;
        w != null && $.set(k, {
          active: E.active,
          id: w
        });
      }
    }, {
      immediate: !0
    }), M(() => st(
      l.value,
      C.value,
      c().flowMode ?? !1,
      O
    ), () => {
      const k = g.value;
      k && Se(
        k,
        l.value,
        C.value,
        c().disableTransform ?? !1,
        c().flowMode ?? !1,
        s,
        O
      );
    }, {
      immediate: !0
    }), Xe(
      f,
      g,
      X,
      S
    )));
    U.mount(), J.set(e, {
      binding: l,
      scope: u,
      options: f,
      callbacks: S,
      el: g,
      controller: U,
      restoreStyles: s
    });
  }
  const Ue = {
    mounted(e, t) {
      if (!z.value)
        return;
      const r = ye(e);
      de(e, t.value, we(t.value, O), r);
    },
    updated(e, t) {
      if (!z.value)
        return;
      const r = J.get(e), s = we(t.value, O);
      if (!r) {
        const l = ye(e);
        de(e, t.value, s, l);
        return;
      }
      const u = r.binding.value !== t.value;
      u && (r.binding.value = t.value), ot(r.options, s), lt(r.callbacks, s), r.el.value !== e && (r.el.value = e), u && Se(
        e,
        t.value,
        C.value,
        c().disableTransform ?? !1,
        c().flowMode ?? !1,
        r.restoreStyles,
        O
      );
    },
    unmounted(e) {
      const t = J.get(e);
      t && (t.controller.unmount(), t.scope.stop(), Te(e, t.restoreStyles), $.delete(e), J.delete(e));
    }
  };
  function H(e = !1) {
    z.value && ((e || F.value) && (v.sizes = {}), d.emit("vscroll:update", { force: !0 }));
  }
  function $e(e, t) {
    T.scrollToItem(e, t);
  }
  function me(e) {
    const t = c();
    return v.sizes = he(e, P.value, F.value ? null : t.keyField), T.restoreCache(e);
  }
  function He(e, t) {
    const r = c(), s = t ?? P.value.indexOf(e), u = F.value ? s : Ie(e, s >= 0 ? s : t, r.keyField);
    return v.sizes[u] || 0;
  }
  function Le() {
    if (!z.value)
      return;
    const e = b.value;
    e && (p || (p = !0, ie(() => {
      e.scrollTop = e.scrollHeight + 5e3;
      const t = () => {
        e.scrollTop = e.scrollHeight + 5e3, W(() => {
          e.scrollTop = e.scrollHeight + 5e3, o === 0 ? p = !1 : W(t);
        });
      };
      W(t);
    })));
  }
  function Q() {
    m && !A && _();
  }
  return M(() => P.value.slice(), (e, t) => {
    if (!z.value)
      return;
    const r = c(), s = F.value ? null : r.keyField, u = V(e, s);
    if (r.shift) {
      const l = t ? V(t, s) : x, f = Je(l, u);
      f > 0 ? (We(l), m && (m.pendingKeys = new Set(u.slice(0, f)), m.stableFrames = 0, ie(() => {
        m && te();
      }))) : _();
    } else
      _();
    x = u, H();
  }, { flush: "sync" }), M(() => c().cache, (e) => {
    z.value && e && me(e);
  }), M(() => c().keyField, (e) => {
    z.value && (v.keyField = e, x = V(P.value, F.value ? null : e), _(), H(!0));
  }, { flush: "sync" }), M(F, (e) => {
    v.simpleArray = e;
  }, { immediate: !0 }), M(() => c().direction, () => {
    z.value && (_(), H(!0));
  }), M(b, (e, t) => {
    t == null || t.removeEventListener("scroll", Q), z.value && (e == null || e.addEventListener("scroll", Q));
  }, {
    immediate: !0
  }), M(z, (e) => {
    const t = b.value;
    e ? (t == null || t.addEventListener("scroll", Q), H()) : (t == null || t.removeEventListener("scroll", Q), ee(), se(), _());
  }), M(oe, () => {
    if (!z.value)
      return;
    const e = b.value;
    if (!e)
      return;
    if (m) {
      te();
      return;
    }
    const t = y;
    if (y = null, !t)
      return;
    const r = N.value.findIndex((l) => l.id === t.key);
    if (r === -1)
      return;
    const s = T.getItemOffset(r) + t.offset;
    s - e.scrollTop !== 0 && (e.scrollTop = s);
  }, { flush: "post" }), Ne(() => {
    v.active = !0;
  }), Qe(() => {
    v.active = !1;
  }), qe(() => {
    var e;
    ee(), se(), (e = b.value) == null || e.removeEventListener("scroll", Q), d.clear();
  }), {
    vscrollData: v,
    itemsWithSize: N,
    resizeObserver: R,
    measurementContext: X,
    vDynamicScrollerItem: Ue,
    ...T,
    pool: Z,
    visiblePool: Ee,
    simpleArray: F,
    forceUpdate: H,
    scrollToItem: $e,
    restoreCache: me,
    getItemSize: He,
    getViewStyle: Ce,
    scrollToBottom: Le,
    onScrollerResize: le,
    onScrollerVisible: ce
  };
}
export {
  mt as useDynamicScroller
};
