// @vitejs/devtools/dist/cli-commands.js
import { n as logger } from "./diagnostics-Bsp9dG9i.js";
import { normalizeHttpServerUrl } from "devframe/node";
import { colors } from "devframe/utils/colors";
import { open } from "devframe/utils/open";
import { resolve } from "pathe";
import { DEVTOOLS_MOUNT_PATH } from "@vitejs/devtools-kit/constants";
//#region src/node/cli-commands.ts
async function start(options) {
	const { host } = options;
	const { getPort } = await import("./dist-BpS-1FhP.js").then((n) => n.t);
	const port = await getPort({
		host,
		port: options.port == null ? void 0 : +options.port,
		portRange: [9999, 15e3]
	});
	const { startStandaloneDevTools } = await import("./standalone-BB0mAs-q.js");
	const { createDevToolsMiddleware } = await import("./server-C2pAbK9p.js").then((n) => n.n);
	const devtools = await startStandaloneDevTools({ cwd: options.root });
	const { h3 } = await createDevToolsMiddleware({
		cwd: devtools.config.root,
		websocket: {
			host,
			https: false
		},
		context: devtools.context
	});
	const { createServer } = await import("node:http");
	const { createApp, eventHandler, sendRedirect, toNodeListener } = await import("h3");
	const { serveStaticHandler } = await import("devframe/utils/serve-static");
	const app = createApp();
	for (const { baseUrl, distDir } of devtools.context.views.buildStaticDirs) app.use(baseUrl, serveStaticHandler(distDir));
	app.use(DEVTOOLS_MOUNT_PATH, h3.handler);
	app.use("/", eventHandler(async (event) => {
		if (event.node.req.url === "/") return sendRedirect(event, DEVTOOLS_MOUNT_PATH);
	}));
	createServer(toNodeListener(app)).listen(port, host, async () => {
		const url = normalizeHttpServerUrl(host, port);
		console.log(colors.green`${"⬢"} Vite DevTools started at`, colors.green(url), "\n");
		if (options.open) await open(url);
	});
}
async function build(options) {
	console.log(colors.cyan`${"⬢"} Building static Vite DevTools...`);
	const { startStandaloneDevTools } = await import("./standalone-BB0mAs-q.js");
	const devtools = await startStandaloneDevTools({
		cwd: options.root,
		config: options.config
	});
	const outDir = resolve(devtools.config.root, options.outDir);
	const { buildStaticDevTools } = await import("./build-static-C-XF3ERp.js");
	await buildStaticDevTools({
		context: devtools.context,
		outDir
	});
	logger.DTK0010().log();
}
//#endregion
export { build, start };
