// @vitejs/devtools/dist/integration.js
//#region src/node/plugins/integration.ts
function getDevToolsEnvironments(config) {
	const environmentNames = config.devtools.config.environments ?? Object.keys(config.environments);
	const environments = [];
	for (const environmentName of environmentNames) {
		const environment = config.environments[environmentName];
		if (environment) environments.push(environment);
	}
	return environments;
}
async function runDevTools$1(builder) {
	const config = builder.config;
	for (const _environment of getDevToolsEnvironments(config)) try {
		const { start } = await import("@vitejs/devtools/cli-commands");
		await start(config.devtools.config);
	} catch (error) {
		config.logger.error(`Failed to run Vite DevTools: ${error?.message || error?.stack || error}`, { error });
	}
}
function DevToolsIntegration$1(_options) {
	return {
		name: "vite:devtools:integration",
		apply: "build",
		configResolved: {
			order: "post",
			handler(config) {
				for (const environment of getDevToolsEnvironments(config)) environment.build.rolldownOptions.devtools ??= {};
			}
		}
	};
}
//#endregion
//#region src/integration.ts
function DevToolsIntegration(options) {
	return DevToolsIntegration$1(options);
}
function runDevTools(builder) {
	return runDevTools$1(builder);
}
//#endregion
export { DevToolsIntegration, runDevTools };
